// alcance/dist/collection/components/alc-tab-button/stories/alc-tab-button.stories.js
// @ts-ignore
import { argTypes } from "./alc-tab-button.args";
import { render as renderArgs } from "../../../stories/functions/components.renderArgs";
import getDefaultArgs from "../../../stories/functions/components.defaultArgs";
export default {
    title: 'Componentes Alfa/Tab Button',
    name: 'Tab Button',
    component: 'alc-tab-button',
    tags: ['alfa'],
    argTypes: argTypes,
};
const defaultArgs = getDefaultArgs(argTypes);
// ESTRUTURA BÁSICA PARA CRIAR EXEMPLOS DO COMPONENTE
// PARA NOVOS EXEMPLOS, REPLIQUE O CÓDIGO ABAIXO E SUBSTITUA O 'Basico' POR OUTRO TIPO DE EXEMPLO: 'Avancado'
export const Basico = {
    name: 'Básico',
    args: Object.assign(Object.assign({}, defaultArgs), {
        // 'selected': ,
        'tab': 'tab-1'
    }),
    render: args => `
<alc-tabs>
  <alc-tab-button slot="button" ${renderArgs(argTypes, args)}>Aba 1</alc-tab-button>
  <alc-tab label="Aba 1" tab="tab-1">
    <p>Conteúdo da Aba 1</p>
  </alc-tab>
</alc-tabs>
  `,
};
export const Selecionado = {
    name: 'Selecionado',
    args: Object.assign(Object.assign({}, defaultArgs), { 'selected': true, 'tab': 'tab-2' }),
    render: args => `
<alc-tabs ${args.selected ? `selected="${args.tab}"` : ''}>
  <alc-tab-button slot="button" tab='tab-1' ${args.tab === 'tab-1' ? 'selected="true"' : ''}>Aba 1</alc-tab-button>
  <alc-tab-button slot="button" tab='tab-2' ${args.tab === 'tab-2' ? 'selected="true"' : ''}>Aba 2</alc-tab-button>
  <alc-tab-button slot="button" tab='tab-3' ${args.tab === 'tab-3' ? 'selected="true"' : ''}>Aba 3</alc-tab-button>
  
  <alc-tab label="Aba 1" tab="tab-1" ${args.tab === 'tab-1' ? 'selected="true"' : ''}>
    <p>Conteúdo da Aba 1</p>
  </alc-tab>
  <alc-tab label="Aba 2" tab="tab-2" ${args.tab === 'tab-2' ? 'selected="true"' : ''}>
    <p>Conteúdo da Aba 2</p>
  </alc-tab>
  <alc-tab label="Aba 3" tab="tab-3" ${args.tab === 'tab-3' ? 'selected="true"' : ''}>
    <p>Conteúdo da Aba 3</p>
  </alc-tab>
</alc-tabs>
  `,
};
