// alcance/dist/collection/components/alc-user/alc-user.js
import { h } from "@stencil/core";
/**
 * @slot DEFAULT - Informação adicional do usuário, não estruturada pelo componente, pode ser colocada aqui. Será exibida imediatamente antes do link "Sair".
 */
export class AlcUser {
    constructor() {
        /**
         * Nome do usuário.
         */
        this.name = 'Usuário';
        /**
         * Número de matrícula ou ponto do usuário.
         */
        this.registrationNumber = '';
        /**
         * URL que efetua o logout do usuário.
         */
        this.logoutUrl = '#';
        /**
         * Imagem do usuário. Pode ser qualquer valor válido para o atributo `src` da tag `img` do HTML.
         */
        this.imgSrc = '';
    }
    async handleLogout(event) {
        const allowLogout = this.alcLogout.emit();
        if (allowLogout.defaultPrevented) {
            event.preventDefault();
        }
    }
    render() {
        return (h("div", { key: 'b880f04f7c16cd12bbff75f682e2027df7aaf0e0', class: "alc-user" }, this.imgSrc ? (h("img", { class: "alc-user__image", src: this.imgSrc, alt: "" })) : (h("alc-icon", { class: "alc-user__image alc-user__image--icon", name: "person-circle", label: "" })), h("div", { key: '8413b4f7e3df2f870b96dc7c46e3e37a62d84de8' }, h("div", { key: '6aee9461e654e08b0e47255c11df6ff4c92e61d0', class: "alc-user__name" }, this.name), h("div", { key: '344f61f96eacf268712084b4990384133d37e494' }, this.registrationNumber), h("slot", { key: '57b56ec0b6450763e1adab33ba6c69abb8ef9d15' }), h("div", { key: '9b10ea4a41c5bd4d0c66effdd0d4ca308ad03706' }, h("alc-icon", { key: '760138d17d0de1b12bf0757e7de19d99cc903059', name: "box-arrow-right", label: "" }), h("a", { key: '9733df1f115340fe20a43165ae8ab935d6ca91a1', class: "alc-link ml-2", href: this.logoutUrl, onClick: (event) => this.handleLogout(event) }, "Sair")))));
    }
    static get is() { return "alc-user"; }
    static get originalStyleUrls() {
        return {
            "$": ["alc-user.css"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["alc-user.css"]
        };
    }
    static get properties() {
        return {
            "name": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Nome do usu\u00E1rio."
                },
                "getter": false,
                "setter": false,
                "attribute": "name",
                "reflect": true,
                "defaultValue": "'Usu\u00E1rio'"
            },
            "registrationNumber": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "N\u00FAmero de matr\u00EDcula ou ponto do usu\u00E1rio."
                },
                "getter": false,
                "setter": false,
                "attribute": "registration-number",
                "reflect": true,
                "defaultValue": "''"
            },
            "logoutUrl": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "URL que efetua o logout do usu\u00E1rio."
                },
                "getter": false,
                "setter": false,
                "attribute": "logout-url",
                "reflect": true,
                "defaultValue": "'#'"
            },
            "imgSrc": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Imagem do usu\u00E1rio. Pode ser qualquer valor v\u00E1lido para o atributo `src` da tag `img` do HTML."
                },
                "getter": false,
                "setter": false,
                "attribute": "img-src",
                "reflect": true,
                "defaultValue": "''"
            }
        };
    }
    static get events() {
        return [{
                "method": "alcLogout",
                "name": "alc-logout",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Evento disparado quando o link \"Sair\" \u00E9 acionado. Se cancelado, n\u00E3o navegar\u00E1 para a URL de logout."
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
}
