// alcance/dist/cjs/alc-view-more.cjs.entry.js
'use strict';

var index = require('./index-BOXTLnv4.js');
var getUniqueId = require('./getUniqueId-DF3jJ4DE.js');

const alcViewMoreCss = "alc-view-more{display:block}.alc-view-more__content{position:relative;display:-ms-flexbox;display:flex;max-height:var(--min-height);-ms-flex-direction:column;flex-direction:column;-ms-flex-pack:justify;justify-content:space-between;overflow:hidden;-webkit-transition-property:all;transition-property:all;-webkit-transition-duration:500ms;transition-duration:500ms;-webkit-transition-timing-function:cubic-bezier(0.4, 0, 0.2, 1);transition-timing-function:cubic-bezier(0.4, 0, 0.2, 1)}.alc-view-more__content--opened{max-height:var(--max-height)}.alc-view-more__content::before{position:absolute;bottom:0px;right:0px;left:0px;height:4rem;background-image:-webkit-gradient(linear, left top, left bottom, from(var(--tw-gradient-stops)));background-image:linear-gradient(to bottom, var(--tw-gradient-stops));--tw-gradient-from:transparent var(--tw-gradient-from-position);--tw-gradient-to:rgb(0 0 0 / 0) var(--tw-gradient-to-position);--tw-gradient-stops:var(--tw-gradient-from), var(--tw-gradient-to);--tw-gradient-to:var(--bg-color) var(--tw-gradient-to-position);--tw-content:\"\";content:var(--tw-content)}.alc-view-more__content--opened::before{display:none}.alc-view-more__toggle{display:-ms-flexbox;display:flex;width:100%}";

const AlcViewMore = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /**
         * Nome da ação para revelar o conteúdo.
         */
        this.toggleTextOpen = 'Veja mais...';
        /**
         * Nome da ação para esconder o conteúdo.
         */
        this.toggleTextClose = 'Fechar';
        /**
         * Alinhamento do botão de ação.
         */
        this.togglePosition = 'center';
        /**
         * Classes CSS para serem aplicadas ao texto da ação.
         */
        this.textClass = '';
        /**
         * Tamanho mínimo do conteúdo a ser exibido.
         * Pode ser em px, rem ou em: 200px; 20rem; 10em;
         */
        this.minHeight = '8em';
        /**
         * Nome da variável CSS correspondente à cor de fundo em que o componente está inserido.
         * Aplica-se ao degradê do corte do texto.
         */
        this.bgColorVar = '--alc-color-surface-primary';
        this.contentId = getUniqueId.getUniqueId();
    }
    toggle(e) {
        e.preventDefault();
        this.contentExpanded = !this.contentExpanded;
    }
    setPosition() {
        if (this.togglePosition === 'right') {
            return 'justify-end';
        }
        if (this.togglePosition === 'left') {
            return 'justify-start';
        }
        return 'justify-center';
    }
    getContentCssClasses() {
        let classes = `alc-view-more__content ${this.textClass}`;
        if (this.contentExpanded) {
            classes = `${classes} alc-view-more__content--opened`;
        }
        return classes;
    }
    componentWillLoad() {
        this.contentExpanded = false;
    }
    componentDidRender() {
        let content = this.el.querySelector('.alc-view-more__content');
        let maxH = content.scrollHeight;
        content.style.setProperty('--min-height', this.minHeight);
        content.style.setProperty('--max-height', maxH + 'px');
        content.style.setProperty('--bg-color', `var(${this.bgColorVar})`);
    }
    render() {
        return (index.h(index.Host, { key: '6ed1099e4c471134cf475145ba8a7df40866e139' }, index.h("div", { key: 'cdcabe99fd1fa7185cc590e41eabdc68d5329d09', id: this.contentId, class: this.getContentCssClasses() }, index.h("slot", { key: '784a5200aadfa30cf70365c147129b97fe48edea' })), index.h("div", { key: 'f67e9efef0ffefafa163244bf01e4a65cefd996d', class: 'alc-view-more__toggle ' + (this.setPosition()) }, index.h("a", { key: '14f59528e78e47e0093ad877c9e436c5e2c9dfb1', href: '#' + this.contentId, onClick: e => { this.toggle(e); }, "aria-expanded": this.contentExpanded.toString(), "aria-controls": this.contentId, role: "button" }, this.contentExpanded ? this.toggleTextClose : this.toggleTextOpen))));
    }
    get el() { return index.getElement(this); }
};
AlcViewMore.style = alcViewMoreCss;

exports.alc_view_more = AlcViewMore;
