// alcance/dist/cjs/alc-event-lab.cjs.entry.js
'use strict';

var index = require('./index-BOXTLnv4.js');
var logger = require('./logger-D5AeZhi-.js');

const alcEventLabCss = ":host{display:block}";

const AlcEventLab = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.alcChange = index.createEvent(this, "alc-change", 7);
        this.alcAfterChange = index.createEvent(this, "alc-after-change", 6);
        this.underlined = 1;
    }
    async changePage(pageNumber) {
        this.requestChangeTo(pageNumber);
    }
    requestChangeTo(newButton) {
        const currentButton = this.underlined;
        logger.logger.debug('RequestChangeTo', newButton);
        // Regra interna do componente, que não faz nada se o botão clicado
        // for o que já está sublinhado
        if (newButton === currentButton) {
            logger.logger.debug('Botão já estava sublinhado');
            return;
        }
        // Dispara o evento buttonWillChange.
        const event = this.alcChange.emit({
            from: currentButton,
            to: newButton
        });
        // Verifica se foi prevenido externamente.
        // Se foi, não faz mais nada.
        if (event.defaultPrevented) {
            logger.logger.debug('defaultPrevented');
            return;
        }
        // Registra a alteração solicitada
        if (newButton === 1) {
            this.b2.classList.remove('underline');
            this.b1.classList.add('underline');
        }
        else if (newButton === 2) {
            this.b1.classList.remove('underline');
            this.b2.classList.add('underline');
        }
        this.underlined = newButton;
        // Dispara o evento buttonDidChange.
        this.alcAfterChange.emit({
            from: currentButton,
            to: newButton
        });
    }
    render() {
        this.alcAfterChange.emit({
            from: undefined,
            to: 1
        });
        return (index.h(index.Host, { key: '916e2c701b44dbbaae07001c91a295b0bb28aec1' }, index.h("button", { key: 'd62e5e9f217ef0b46e56556c4703f3cf86795c57', class: "button underline", ref: el => this.b1 = el, onClick: () => this.requestChangeTo(1) }, "1"), index.h("button", { key: '177c7edc53e2ff0aef977b41a598860d6e8b3812', class: "button", ref: el => this.b2 = el, onClick: () => this.requestChangeTo(2) }, "2"), index.h("slot", { key: '1b6a3447a5b462ea691f67dfeff86d8ecbcacc42' })));
    }
};
AlcEventLab.style = alcEventLabCss;

exports.alc_event_lab = AlcEventLab;
