// alcance/dist/cjs/alc-user-menu.cjs.entry.js
'use strict';

var index = require('./index-BOXTLnv4.js');
var logger = require('./logger-D5AeZhi-.js');

const alcUserMenuCss = "alc-user-menu{--avatar-size:2.5rem;--avatar-size-mobile:4rem;--border-width:2px}.alc-user-menu__image{margin-right:var(--alc-spacing-03);height:var(--avatar-size);width:var(--avatar-size);border-radius:9999px;border-width:calc(var(--border-width) * 1);--tw-border-opacity:1;border-color:rgb(0 104 125 / var(--tw-border-opacity));-o-object-fit:cover;object-fit:cover}.alc-user-menu__image:is([data-alc-theme=\"dark\"] *){--tw-border-opacity:1;border-color:rgb(79 158 153 / var(--tw-border-opacity))}.alc-user-menu__icon{margin-right:var(--alc-spacing-03);border-radius:9999px;border-width:calc(var(--border-width) * 1);--tw-border-opacity:1;border-color:rgb(0 104 125 / var(--tw-border-opacity));-o-object-fit:cover;object-fit:cover;font-size:calc(var(--avatar-size) - 2 * var(--border-width));line-height:1}.alc-user-menu__icon:is([data-alc-theme=\"dark\"] *){--tw-border-opacity:1;border-color:rgb(79 158 153 / var(--tw-border-opacity))}.alc-user-menu--mobile{display:-ms-flexbox;display:flex;-ms-flex-direction:column;flex-direction:column}.alc-user-menu--mobile__header{display:-ms-flexbox;display:flex;-ms-flex-align:center;align-items:center;padding-left:var(--alc-spacing-06);padding-right:var(--alc-spacing-09);padding-top:var(--alc-spacing-06)}.alc-user-menu--mobile__header .alc-user-menu__image{margin-right:var(--alc-spacing-05);height:var(--avatar-size-mobile);width:var(--avatar-size-mobile);border-style:none}.alc-user-menu--mobile__header .alc-user-menu__icon{background-color:var(--alc-color-layer-a);font-size:calc(var(--avatar-size-mobile) * 1);line-height:1}.alc-user-menu--mobile .alc-user-menu__name{font-size:1rem;line-height:1.5rem;font-weight:500}.alc-user-menu--desktop{display:-ms-flexbox;display:flex;-ms-flex-align:center;align-items:center;-ms-flex-pack:justify;justify-content:space-between}.alc-user-menu__info--external{margin-right:var(--alc-spacing-03);margin-left:auto;display:-ms-flexbox;display:flex;-ms-flex-direction:column;flex-direction:column;text-align:right}.alc-user-menu__info--internal{padding-left:var(--alc-spacing-05);padding-right:var(--alc-spacing-05);padding-bottom:var(--alc-spacing-03);padding-top:var(--alc-spacing-05);color:var(--alc-color-text-primary)}.alc-user-menu__info--internal:empty{padding:0px}.alc-user-menu__info{margin-right:var(--alc-spacing-03);display:-ms-flexbox;display:flex;-ms-flex-direction:column;flex-direction:column}.alc-user-menu__primary{-ms-flex-align:center;align-items:center;gap:var(--alc-spacing-02)}.alc-user-menu__name{font-size:1rem;line-height:var(--alc-font-line-height-base);font-weight:500}.alc-user-menu__reg{margin-left:var(--alc-spacing-02);font-weight:400}.alc-user-menu--mobile__reg{margin-left:var(--alc-spacing-02);font-weight:400}.alc-popup__content{max-width:16rem}.alc-user-menu__trigger{display:-ms-flexbox;display:flex;cursor:pointer;-ms-flex-align:center;align-items:center;border-style:none;background-color:transparent;padding:0px}.alc-user-menu__dropdown alc-popup{display:-ms-flexbox;display:flex}.alc-user-menu__secondary{font-size:0.875rem;line-height:1rem;font-weight:400;color:var(--alc-color-text-subtle)}.alc-user-menu__info--external .alc-user-menu__secondary{color:currentColor}.alc-user-menu__logout-icon{margin-right:var(--alc-spacing-03)}.alc-user-menu__info--separator{border-bottom-width:1px;border-color:var(--alc-color-border)}";

const AlcUserMenu = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.alcLogout = index.createEvent(this, "alc-logout", 7);
        /**
         * Define a variação do menu de usuário. Geralmente, não é recomendado alterar esse valor.
         */
        this.variation = 'desktop';
        /**
         * Nome do usuário.
         */
        this.name = 'Usuário';
        /**
         * Número de matrícula ou ponto do usuário.
         */
        this.registrationNumber = '';
        /**
         * URL da imagem do usuário. Se não for fornecida, será exibido um ícone padrão.
         */
        this.imgSrc = '';
        this.dropdownOpen = false;
        this.handleLogout = (e) => {
            const { defaultPrevented } = this.alcLogout.emit();
            if (defaultPrevented) {
                e.preventDefault();
            }
        };
    }
    componentDidLoad() {
        if (this.variation === 'desktop') {
            this.externalContainer = this.host.querySelector('.alc-user-menu__info--external');
            this.internalContainer = this.host.querySelector('.alc-user-menu__info--internal');
            this.moveInfo();
        }
    }
    handleResize() {
        if (this.variation === 'desktop') {
            this.moveInfo();
        }
    }
    moveInfo() {
        const largura = window.innerWidth;
        // < 1200px é compacto, >= 1200px é expandido
        if (largura < 1200) {
            if (this.externalContainer && this.internalContainer && this.internalContainer.children.length === 0) {
                while (this.externalContainer.firstChild) {
                    this.internalContainer.appendChild(this.externalContainer.firstChild);
                    this.internalContainer.classList.add('alc-user-menu__info--separator');
                }
            }
        }
        else {
            if (this.externalContainer && this.internalContainer && this.externalContainer.children.length === 0) {
                while (this.internalContainer.firstChild) {
                    this.externalContainer.appendChild(this.internalContainer.firstChild);
                    this.internalContainer.classList.remove('alc-user-menu__info--separator');
                }
            }
        }
    }
    componentDidRender() {
        if (this.variation === 'mobile') {
            this.mobileMenuContainer.insertAdjacentElement('beforeend', this.menu);
            this.mobileSlotContainer.insertAdjacentElement('beforeend', this.defaultSlot);
        }
        else {
            this.desktopMenuContainer.insertAdjacentElement('beforeend', this.menu);
            this.desktopSlotContainer.insertAdjacentElement('beforeend', this.defaultSlot);
        }
    }
    renderDesktop() {
        return (index.h("div", { class: {
                "alc-user-menu--desktop": true,
                'hidden': this.variation === 'mobile'
            } }, index.h("div", { class: "alc-user-menu__info--external" }, index.h("div", null, index.h("div", { class: "alc-user-menu__primary" }, index.h("span", { class: "alc-user-menu__name" }, this.name), this.registrationNumber && index.h("span", { class: "alc-user-menu__reg" }, "(", this.registrationNumber, ")")), index.h("div", { class: "alc-user-menu__secondary", ref: el => this.desktopSlotContainer = el }))), index.h("alc-dropdown", { class: "alc-user-menu__dropdown" }, index.h("button", { slot: "trigger", class: "alc-user-menu__trigger", "aria-label": "Abrir menu do usu\u00E1rio" }, this.imgSrc ? (index.h("img", { class: "alc-user-menu__image", src: this.imgSrc, alt: this.name })) : (index.h("alc-icon", { class: "alc-user-menu__image alc-user-menu__icon", name: "person-fill", label: this.name })), index.h("alc-icon", { icon: "chevron-down", label: "" })), index.h("div", { class: "alc-user-menu__info--internal" }), index.h("div", { ref: el => this.desktopMenuContainer = el }))));
    }
    renderMobile() {
        return (index.h("div", { class: {
                "alc-user-menu--mobile": true,
                'hidden': this.variation === 'desktop'
            } }, index.h("div", { class: "alc-user-menu--mobile__header" }, this.imgSrc ? (index.h("img", { class: "alc-user-menu__image", src: this.imgSrc, alt: this.name })) : (index.h("alc-icon", { class: "alc-user-menu__image alc-user-menu__icon", name: "person-fill", label: this.name })), index.h("div", { class: "alc-user-menu--mobile__info" }, index.h("div", { class: "alc-user-menu__name" }, this.name, this.registrationNumber && index.h("span", { class: "alc-user-menu--mobile__reg" }, "(", this.registrationNumber, ")")), index.h("div", { class: "alc-user-menu__secondary", ref: el => this.mobileSlotContainer = el }))), index.h("alc-expander", { label: "Menu do usu\u00E1rio", open: false, "hide-label": "true" }, index.h("div", null, index.h("div", { ref: el => this.mobileMenuContainer = el })))));
    }
    render() {
        var _a;
        (_a = this.logoutUrl) !== null && _a !== void 0 ? _a : logger.logger.report('logoutUrl', this.host.tagName.toLowerCase(), this.host);
        return (index.h(index.Host, { key: '1df7026fa5897691438a13aebfb38ad075f942f0' }, index.h("alc-menu", { key: '113f70bfb848564bf76a688356a458042e5e4cc0', ref: el => this.menu = el }, index.h("slot", { key: '13bdfda9dbb3a473380af99c6f47dde82f81f065', name: "actions" }), index.h("alc-menu-item-theme", { key: '5dcb6428c37d40843c29d77c5d3c164516646c0a' }), index.h("alc-menu-link", { key: 'dc29f0f8eec9b925e2f99330113f6b5a26d843b6' }, index.h("a", { key: '93b72f1a292ed1a45250aa7b515b40f2480c05a2', href: this.logoutUrl, onClick: this.handleLogout }, index.h("alc-icon", { key: '50c85a0422c3d2d87b340309917f12c0dedf4a37', name: "box-arrow-right", label: "", class: "alc-user-menu__logout-icon" }), "Sair"))), index.h("div", { key: '9b8f2386263e18f980625e276ba8eb72a152e28d', ref: el => this.defaultSlot = el }, index.h("slot", { key: '99a076179acea8e92cc4bf7f825593fc2dede355' })), [
            this.renderMobile(),
            this.renderDesktop()
        ]));
    }
    get host() { return index.getElement(this); }
};
AlcUserMenu.style = alcUserMenuCss;

exports.alc_user_menu = AlcUserMenu;
