// alcance/dist/cjs/alc-header.cjs.entry.js
'use strict';

var index = require('./index-BOXTLnv4.js');
var tailwindScreens = require('./tailwind-screens-B7NOrE0I.js');
var logger = require('./logger-D5AeZhi-.js');
var getUniqueId = require('./getUniqueId-DF3jJ4DE.js');
var throttle = require('./throttle-DAOUKsy2.js');

const alcHeaderCss = "[data-alc-theme=\"light\"] alc-header{--bg-color-gradient-from:#40807e;--bg-color-gradient-to:#074b69;--bg-color-logo:#2a646d}[data-alc-theme=\"dark\"] alc-header{--bg-color-gradient-from:#203131;--bg-color-gradient-to:#1c1d1f;--bg-color-logo:#172222}alc-header{position:-webkit-sticky;position:sticky;top:0px;z-index:20;display:block}.alc-header{-webkit-transition:-webkit-transform 0.3s ease;transition:-webkit-transform 0.3s ease;transition:transform 0.3s ease;transition:transform 0.3s ease, -webkit-transform 0.3s ease}.alc-header--scroll-down{--tw-translate-y:-100%;-webkit-transform:translate(var(--tw-translate-x), var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));transform:translate(var(--tw-translate-x), var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y))}@media (min-width: 768px){.alc-header--scroll-down{--tw-translate-y:0px;-webkit-transform:translate(var(--tw-translate-x), var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));transform:translate(var(--tw-translate-x), var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y))}}.alc-header__header{display:block;width:100%}@media (min-width: 768px){.alc-header__header{z-index:1020}}.alc-header__id{display:-ms-flexbox;display:flex;gap:var(--alc-spacing-05)}.alc-header__container{display:-ms-flexbox;display:flex;-ms-flex-align:center;align-items:center;-ms-flex-pack:justify;justify-content:space-between;gap:var(--alc-spacing-05);background-image:-webkit-gradient(linear, left top, right top, from(var(--tw-gradient-stops)));background-image:linear-gradient(to right, var(--tw-gradient-stops));--tw-gradient-from-position:0%;--tw-gradient-to-position:100%;padding:var(--alc-spacing-05);--tw-gradient-from:var(--bg-color-gradient-from) var(--tw-gradient-from-position);--tw-gradient-to:rgb(255 255 255 / 0) var(--tw-gradient-to-position);--tw-gradient-stops:var(--tw-gradient-from), var(--tw-gradient-to);--tw-gradient-to:var(--bg-color-gradient-to) var(--tw-gradient-to-position)}.alc-header__logo{height:2.25rem;width:2.25rem}@media (min-width: 992px){.alc-header__logo{height:2.5rem;width:2.5rem}}.alc-header__panel{display:-ms-flexbox;display:flex;-ms-flex-positive:1;flex-grow:1;-ms-flex-align:center;align-items:center}.alc-header__support-wrapper{-ms-flex-positive:1;flex-grow:1;white-space:nowrap;text-align:right}.alc-header__support{display:inline-block}.alc-header__menu-separator{font-size:0.75rem;line-height:1rem;font-weight:700;text-transform:uppercase;letter-spacing:0.1em;color:var(--alc-color-text-subtle);padding-top:var(--alc-spacing-04);padding-bottom:var(--alc-spacing-04);padding-left:var(--alc-spacing-05);padding-right:var(--alc-spacing-05);border-top-width:1px;border-color:var(--alc-color-border)}.alc-header__user{--padding:var(--alc-spacing-05);--margin:var(--alc-spacing-05);--border-width:1px;--blank-space:calc( var(--margin) + var(--border-width) + var(--padding));margin-left:var(--blank-space);display:none;--tw-text-opacity:1;color:rgb(255 255 255 / var(--tw-text-opacity))}@media (min-width: 768px){.alc-header__user{display:block}}.alc-header__user--separator{margin-left:var(--margin);border-left-width:calc(var(--border-width) * 1);--tw-border-opacity:1;border-left-color:rgb(255 255 255 / var(--tw-border-opacity));padding-left:var(--padding)}@media (min-width: 768px){.alc-header__drawer-nav{display:none}}";

// Pega o valor definido da configuração do tailwind e remove o 'px'
const MOBILE_BREAKPOINT = Number(tailwindScreens.screens.sm.split('px')[0]);
const TABLET_BREAKPOINT = Number(tailwindScreens.screens.md.split('px')[0]);
const AlcHeader = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.alcHome = index.createEvent(this, "alc-home", 7);
        this.supportItems = [];
        this.availableWidth = 0;
        this.requiredWidth = 0;
        this.scrollTop = 0;
        this.scrollDirection = null;
        this.resizeThrottle = throttle.throttle(this.resize, 300);
        this.isDrawer = false;
        this.isSupportDropdown = false;
        this.needSeparator = false;
        this.isMenuCreated = false;
    }
    handleScroll() {
        if (window.innerWidth >= TABLET_BREAKPOINT)
            return;
        const currentScroll = window.pageYOffset || document.documentElement.scrollTop;
        let currentDirection = this.scrollDirection;
        if (currentScroll > this.scrollTop && currentDirection !== 'down' && currentScroll > this.el.clientHeight / 2) {
            // Scroll descendo, esconde
            this.wrapper.classList.add('alc-header--scroll-down');
            currentDirection = 'down';
        }
        if (currentScroll < this.scrollTop && currentDirection !== 'up') {
            // Scroll subindo, mostra
            currentDirection = 'up';
            this.wrapper.classList.remove('alc-header--scroll-down');
        }
        this.scrollTop = currentScroll <= 0 ? 0 : currentScroll;
        this.scrollDirection = currentDirection;
    }
    handleNavContent(event) {
        var _a, _b;
        const { navEl, isMobile } = event.detail;
        // pega o drawer se ainda não pegou
        const drawer = (_a = this.drawer) !== null && _a !== void 0 ? _a : this.el.querySelector('alc-drawer');
        // pega o nav-panel se ainda não pegou
        const navPanel = (_b = this.navPanel) !== null && _b !== void 0 ? _b : document.querySelector('alc-nav-panel');
        if (!drawer || !navPanel) {
            // ainda não inicializou completamente, sai sem fazer nada
            return;
        }
        // PROCURA O BODY DO DRAWER, OU USA O PRÓPRIO drawer COMO BACKUP
        const drawerBody = drawer.querySelector('.alc-drawer__header + div') || drawer;
        const panelContent = navPanel.querySelector('.alc-nav-panel__content');
        if (!drawerBody || !panelContent) {
            return;
        }
        if (isMobile) {
            // MODO MOBILE
            console.log("É Mobile!");
            if (panelContent.contains(navEl)) {
                panelContent.removeChild(navEl);
            }
            if (!drawerBody.contains(navEl)) {
                drawerBody.appendChild(navEl);
            }
            navPanel.classList.add('alc-nav-panel--hidden');
        }
        else {
            // MODO DESKTOP
            if (drawerBody.contains(navEl)) {
                drawerBody.removeChild(navEl);
            }
            if (!panelContent.contains(navEl)) {
                panelContent.appendChild(navEl);
            }
            navPanel.classList.remove('alc-nav-panel--hidden');
        }
    }
    handleResize() {
        this.resizeThrottle();
    }
    resize() {
        // Se o drawer estiver aberto e a tela for maior que o breakpoint, fecha o drawer.
        if (window.innerWidth >= TABLET_BREAKPOINT && this.drawer.isVisible) {
            this.drawer.hide();
        }
    }
    handleAlcHome(event) {
        event.stopPropagation();
        const homeEvent = this.alcHome.emit();
        if (homeEvent.defaultPrevented) {
            event.preventDefault();
        }
    }
    componentDidRender() {
        // Tabelas de conversão de variante
        const toMenuVariant = {
            link: 'menu-link',
            'menu-link': 'menu-link',
            button: 'menu-item',
            'menu-item': 'menu-item',
            null: 'menu-item',
        };
        const toInlineVariant = {
            link: 'link',
            'menu-link': 'link',
            button: 'button',
            'menu-item': 'button',
            null: 'button',
        };
        // Obtém os elementos do slot support
        if (!this.supportItems.length) {
            const assignedElementsSlotSupport = this.slotSupport.assignedElements();
            if (assignedElementsSlotSupport.length > 0) {
                Array.from(assignedElementsSlotSupport[0].children).forEach(item => {
                    this.supportItems.push(item);
                });
            }
        }
        if (!this.userMenu) {
            this.userMenu = this.slotUser.assignedElements()[0].firstElementChild;
        }
        if (!this.menuSupport) {
            this.menuSupport = document.createElement('div');
        }
        logger.logger.debug('header componentDidRender', this.supportItems, this.slotSupport.assignedElements());
        // Remove filhos de menuSupport
        this.menuSupport.replaceChildren();
        /*
          Dois casos em que o menu de apoio estará "junto" ao menu de usuário:
          1. Quando não houver espaço para os itens de apoio lado a lado no header (isSupportDropdown)
          2. Quando a navegação estiver contida no drawer - padrão para as telas menores (isDrawer)
        */
        if (this.isSupportDropdown || this.isDrawer) {
            logger.logger.debug('header createElement menu', this.supportItems);
            const menu = this.userMenu.querySelector('alc-menu');
            if (this.supportItems.length > 0) {
                // Cria um grupo e um separador dentro dele. O separador dá nome ao grupo.
                const supportGroup = document.createElement('div');
                supportGroup.role = "group";
                const separator = document.createElement('div');
                // @TODO: Estruturar isso depois dentro de alc-menu (ou componentes relacionados)
                separator.className = "alc-header__menu-separator";
                separator.role = "separator";
                separator.innerText = "Menu de apoio";
                separator.id = getUniqueId.getUniqueId();
                supportGroup.appendChild(separator);
                supportGroup.setAttribute('aria-labelledby', separator.id);
                Array.from(this.supportItems).forEach(item => {
                    // @ts-ignore
                    item.variant = toMenuVariant[item.variant];
                    // @ts-ignore
                    supportGroup.appendChild(item);
                });
                logger.logger.debug('header insertElement menu', this.supportItems, supportGroup);
                this.menuSupport.appendChild(supportGroup);
                menu.insertAdjacentElement('beforeend', this.menuSupport);
            }
            // @TODO: Encontrar solução mais robusta do que um timeout.
            setTimeout(() => {
                logger.logger.debug('menu.refresh()');
                menu.refresh();
            }, 100);
        }
        else {
            Array.from(this.supportItems).forEach(item => {
                // @ts-ignore
                item.variant = toInlineVariant[item.variant];
                // @ts-ignore
                this.support.appendChild(item);
            });
        }
        if (this.isDrawer) {
            this.userMenu.variation = 'mobile';
            this.drawer.insertAdjacentElement('beforeend', this.userMenu);
            // @TODO: Aqui criou-se uma dependência da estrutura interna do dropdown
            // a necessidade especificar a classe alc-drawer__header.
            // Isso é um ponto de atenção e que merecer ser revisto para uma solução mais robusta.
            this.drawer.querySelector('.alc-drawer__header + div').insertAdjacentElement('afterbegin', this.userMenu);
        }
        else {
            this.userMenu.variation = 'desktop';
            this.slotUser.assignedElements()[0].insertAdjacentElement('beforeend', this.userMenu);
        }
        logger.logger.debug(this.supportItems);
    }
    componentDidLoad() {
        // ref ao drawer já existente no template
        this.drawer = this.el.querySelector('alc-drawer');
        // nav-panel está em outro lugar do DOM
        this.navPanel = document.querySelector('alc-nav-panel');
        this.observer = new ResizeObserver(() => {
            this.checkLayout();
        });
        // @TODO: Encontrar solução mais robusta do que um timeout.
        setTimeout(() => {
            this.calcRequiredWidth();
            this.observer.observe(this.el);
        }, 100);
    }
    render() {
        var _a, _b;
        (_a = this.name) !== null && _a !== void 0 ? _a : logger.logger.report('name', this.el.tagName.toLowerCase(), this.el);
        (_b = this.homeUrl) !== null && _b !== void 0 ? _b : logger.logger.report('homeUrl', this.el.tagName.toLowerCase(), this.el);
        return (index.h(index.Host, { key: '50843cc351b03605b026104a99b765dbd2ec1af5' }, index.h("div", { key: 'acd386bda99c06051a3ecc8c39c7a311bba075aa', class: "alc-header", ref: (el) => this.wrapper = el }, index.h("header", { key: '8dc133f406d7e46b7ed5c6a43370c9e6195eb24b', class: "alc-header__header" }, index.h("slot", { key: '33803ee86ade7cd59fdf3522e6a52be4f143ecd3' }), index.h("slot", { key: 'a12d423b6d9a76409c8b6407be44c9e38b383823', name: "support", ref: (el) => this.slotSupport = el }), index.h("div", { key: '8f5a09861a39c6554cf9546d0ff148834802a222', class: "alc-header__container" }, index.h("div", { key: '8ab01c8f29de13803764c09f6acfcec29fb54a02', class: "alc-header__id" }, index.h("alc-header-id", { key: 'ae05e593ef6c66b4014b4a4de6472fa13807ed4d', name: this.name, description: this.description, homeUrl: this.homeUrl, "onAlc-home": this.handleAlcHome.bind(this) })), index.h("div", { key: '363e93508d9792a435031c168a1e6447977c1c66', class: "alc-header__panel" }, index.h("div", { key: 'ce81a3325dd43039f552b4a4a672046525bbc06e', class: "alc-header__support-wrapper", ref: el => this.supportWrapper = el }, index.h("div", { key: '06ac79d9b6afaf7019e1aec483b6c439cc7f00e7', class: "alc-header__support", ref: el => this.support = el })), index.h("div", { key: '97879eb8fa630920969b381d3d6c70603efe118e', class: "alc-header__fixed" }, index.h("slot", { key: '18e753581b7bc1550c147c86d82afc1ef62d52a7', name: "fixed", ref: (el) => this.slotFixed = el })), index.h("div", { key: 'bc5a9abe1c62324b5b31f7da8e6ef1f400232e6f', class: "alc-header__drawer-nav" }, index.h("button", { key: '45ccd23d19b2596a114d5a06e1542c4f2c894ab7', class: "alc-header-button", onClick: () => this.drawer.show(), "data-alc-header-drawer-button": true }, index.h("alc-icon", { key: '034076d0807c83526105def7a2a5c0cb65edd68e', icon: "list", label: "Abrir navega\u00E7\u00E3o", class: "alc-header-button__icon" })), index.h("alc-drawer", { key: '30239f69ac0224bea50bf15e9a867c8caaf70940', ref: (el) => this.drawer = el })), index.h("div", { key: '183ef67e4d5b88f4c9b0a627ea37d718e498e38a', class: {
                'alc-header__user': true,
                'alc-header__user--separator': this.needSeparator
            } }, index.h("slot", { key: 'e43720d6466616b229f77420745637b68a63c24d', name: "user", ref: (el) => this.slotUser = el }))))))));
    }
    calcRequiredWidth() {
        // Altera overflow-y para "hidden" para fazer a medição
        this.supportWrapper.style.overflowY = 'hidden';
        this.supportWrapper.style.overflowX = 'clip';
        this.requiredWidth = this.support.scrollWidth;
        this.supportWrapper.style.removeProperty('overflow-x');
        this.supportWrapper.style.removeProperty('overflow-y');
    }
    checkLayout() {
        // Altera overflow-y para "hidden" para fazer a medição
        this.supportWrapper.style.overflowY = 'hidden';
        this.supportWrapper.style.overflowX = 'clip';
        this.availableWidth = this.supportWrapper.offsetWidth;
        this.isSupportDropdown = (this.requiredWidth > this.availableWidth) || (window.innerWidth < MOBILE_BREAKPOINT);
        this.supportWrapper.style.removeProperty('overflow-x');
        this.supportWrapper.style.removeProperty('overflow-y');
        this.isDrawer = window.innerWidth < TABLET_BREAKPOINT;
        this.needSeparator = this.fixedElementsVisible() || (this.requiredWidth > 0 && !this.isSupportDropdown);
    }
    fixedElementsVisible() {
        const slotElement = this.slotFixed.assignedElements()[0];
        // O slot foi usado
        if (slotElement instanceof HTMLElement) {
            // True se está ocupando alguma largura (ou seja, se tem algo visível)
            return (slotElement.offsetWidth > 0);
        }
        return false;
    }
    get el() { return index.getElement(this); }
};
AlcHeader.style = alcHeaderCss;

exports.alc_header = AlcHeader;
