// alcance/dist/cjs/alc-radio.cjs.entry.js
'use strict';

var index = require('./index-BOXTLnv4.js');
var getUniqueId = require('./getUniqueId-DF3jJ4DE.js');
var testAttributes = require('./testAttributes-uTFBrsRA.js');

const alcRadioCss = "[data-alc-theme=\"dark\"] alc-radio{--accent-color:var(--alc-color-action-40);--outline-color:var(--alc-color-action-30);--text-color:var(--alc-color-text-4);--text-color-error:var(--alc-color-error-45)}[data-alc-theme=\"light\"] alc-radio{--accent-color:var(--alc-color-action-60);--outline-color:var(--alc-color-action-60);--text-color:var(--alc-color-text-4);--text-color-error:var(--alc-color-error-50)}alc-radio{display:-ms-flexbox;display:flex;-ms-flex-align:start;align-items:flex-start;gap:0.5rem}alc-radio input[type='radio']{margin-top:1px}alc-radio input[type='radio']:focus{outline-style:solid;outline-width:2px;outline-offset:2px}alc-radio input[type='radio']{accent-color:var(--accent-color)}alc-radio input[type='radio']:focus{outline-color:var(--outline-color)}.alc-radio__label{display:-ms-flexbox;display:flex;-ms-flex-direction:column;flex-direction:column;font-size:0.875rem;line-height:1rem;font-weight:500}.alc-radio__text{font-size:0.875rem;line-height:1rem;font-weight:400;color:var(--text-color)}.alc-radio__text--error{color:var(--text-color-error)}";

const AlcRadio = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.describedBy = new Map();
        this.errorId = '';
        this.hintId = '';
        /**
         * Texto de ajuda.
        */
        this.hint = '';
        /**
         * Texto do label do input. Há opção de adicionar o elemento label pelo slot "label".
        */
        this.label = '';
        /**
         * Texto de mensagem de erro.
        */
        this.errorMsg = '';
    }
    componentWillLoad() {
        // Independentemente de qualquer coisa, reserva ids para esses elementos.
        this.errorId = getUniqueId.getUniqueId();
        this.hintId = getUniqueId.getUniqueId();
        this.input = this.el.querySelector('input[type="radio"]');
        const labelElement = this.el.querySelector('label');
        this.inputId = this.input.id;
        if (!this.inputId) {
            this.inputId = getUniqueId.getUniqueId();
            this.input.setAttribute('id', this.inputId);
        }
        if (labelElement && !labelElement.getAttribute('for')) {
            labelElement.setAttribute('for', this.inputId);
        }
        this.describedBy.set('original', this.input.getAttribute('aria-describedby') || '');
    }
    componentWillRender() {
        if (this.errorMsg) {
            this.describedBy.set('error', this.errorId);
        }
        else {
            this.describedBy.delete('error');
        }
        if (this.hint) {
            this.describedBy.set('hint', this.hintId);
        }
        else {
            this.describedBy.delete('hint');
        }
        let currentDescribedBy = '';
        currentDescribedBy += ` ${this.describedBy.get('original') || ''}`;
        currentDescribedBy += ` ${this.describedBy.get('error') || ''}`;
        currentDescribedBy += ` ${this.describedBy.get('hint') || ''}`;
        this.input.setAttribute('aria-describedby', currentDescribedBy.trim());
    }
    render() {
        return (index.h(index.Host, { key: '7de3e69e23473c75f75b298af398d03a05f75229' }, index.h("slot", { key: '94934d27a296bdee39ef63645250f61bce5f7824' }), index.h("div", { key: '2965ce9a65f9a1912604b47010585142c82dd45f', class: "alc-radio__label" }, index.h("div", { key: '8a088e0c25d1defce1c2bb95fbffdebb3a8bac01' }, this.label
            ?
                index.h("label", Object.assign({ htmlFor: this.inputId }, testAttributes.testAttributes()), this.label)
            :
                index.h("slot", { name: "label" })), this.errorMsg
            ?
                index.h("small", Object.assign({ class: "alc-radio__text alc-radio__text--error", id: this.errorId }, testAttributes.testAttributes()), this.errorMsg)
            :
                null, this.hint
            ?
                index.h("small", Object.assign({ class: "alc-radio__text", id: this.hintId }, testAttributes.testAttributes()), this.hint)
            :
                null)));
    }
    get el() { return index.getElement(this); }
};
AlcRadio.style = alcRadioCss;

exports.alc_radio = AlcRadio;
