// alcance/dist/cjs/alc-breadcrumb-item.cjs.entry.js
'use strict';

var index = require('./index-BOXTLnv4.js');
var testAttributes = require('./testAttributes-uTFBrsRA.js');

const alcBreadcrumbItemCss = "[data-alc-theme=\"dark\"] alc-breadcrumb-item{--outline-color:var(--alc-color-action-30)}[data-alc-theme=\"light\"] alc-breadcrumb-item{--outline-color:var(--alc-color-action-60)}.alc-breadcrumb-item{display:-ms-flexbox;display:flex;-ms-flex-align:baseline;align-items:baseline}.alc-breadcrumb-item__link{display:-ms-flexbox;display:flex;-ms-flex-align:baseline;align-items:baseline;color:var(--alc-color-text-secondary)}.alc-breadcrumb-item__icon{margin-right:0.25rem}.alc-breadcrumb-item__link a{border-radius:0.125rem;outline-color:var(--outline-color)}.alc-breadcrumb-item__separator{margin-left:0.25rem;margin-right:0.25rem}";

const AlcBreadcrumbItem = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /** Indica a url do link, pode-se usar o slot default para adicionar o elemento de navegação */
        this.url = null;
        /** Indica a label do breadcrumb */
        this.label = null;
        /** Indica o nome do icone a ser renderizado */
        this.iconName = null;
        this.isCurrentPage = false;
    }
    componentWillLoad() {
        this.isCurrentPage = !!this.el.ariaCurrent;
    }
    onAriaLabelChange(newValue) {
        this.isCurrentPage = newValue === null ? false : true;
    }
    render() {
        return (index.h(index.Host, { key: 'eb3336d1010473107ba48af621865847c669b7e0', role: "listitem" }, index.h("div", { key: '843d998744324ad3e0e4b38c07db32dda9fff05a', class: 'alc-breadcrumb-item' }, index.h("div", { key: '5a271f68cc38681bab776a0b8ceda5ce53f859f3', class: {
                'alc-breadcrumb-item__link': true,
                'alc-link': !this.isCurrentPage
            } }, this.iconName && (index.h("alc-icon", Object.assign({ key: '2b47b53113ed0e955b4c9b022a9aa7a508a54e02', label: "", name: this.iconName, class: "alc-breadcrumb-item__icon", "aria-hidden": "true" }, testAttributes.testAttributes()))), this.url ? (index.h("a", Object.assign({ href: this.url }, testAttributes.testAttributes()), this.label)) : (index.h("slot", null))), !this.isCurrentPage && (index.h("span", { key: '4d72144f8495b0da57230a7c3af5d403f127b9e8', class: 'alc-breadcrumb-item__separator' }, "/")))));
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "aria-current": ["onAriaLabelChange"]
    }; }
};
AlcBreadcrumbItem.style = alcBreadcrumbItemCss;

exports.alc_breadcrumb_item = AlcBreadcrumbItem;
