// alcance/dist/cjs/alc-user.cjs.entry.js
'use strict';

var index = require('./index-BOXTLnv4.js');

const alcUserCss = ".alc-user{display:-ms-flexbox;display:flex}.alc-user__image{margin-right:1rem;height:3.5rem;width:3.5rem;border-radius:9999px;-o-object-fit:cover;object-fit:cover}.alc-user__image--icon{font-size:3.5rem;line-height:1}.alc-user__name{margin-top:0.5rem;font-size:1.125rem;line-height:1.5rem;font-weight:700}";

const AlcUser = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.alcLogout = index.createEvent(this, "alc-logout", 7);
        /**
         * Nome do usuário.
         */
        this.name = 'Usuário';
        /**
         * Número de matrícula ou ponto do usuário.
         */
        this.registrationNumber = '';
        /**
         * URL que efetua o logout do usuário.
         */
        this.logoutUrl = '#';
        /**
         * Imagem do usuário. Pode ser qualquer valor válido para o atributo `src` da tag `img` do HTML.
         */
        this.imgSrc = '';
    }
    async handleLogout(event) {
        const allowLogout = this.alcLogout.emit();
        if (allowLogout.defaultPrevented) {
            event.preventDefault();
        }
    }
    render() {
        return (index.h("div", { key: 'b880f04f7c16cd12bbff75f682e2027df7aaf0e0', class: "alc-user" }, this.imgSrc ? (index.h("img", { class: "alc-user__image", src: this.imgSrc, alt: "" })) : (index.h("alc-icon", { class: "alc-user__image alc-user__image--icon", name: "person-circle", label: "" })), index.h("div", { key: '8413b4f7e3df2f870b96dc7c46e3e37a62d84de8' }, index.h("div", { key: '6aee9461e654e08b0e47255c11df6ff4c92e61d0', class: "alc-user__name" }, this.name), index.h("div", { key: '344f61f96eacf268712084b4990384133d37e494' }, this.registrationNumber), index.h("slot", { key: '57b56ec0b6450763e1adab33ba6c69abb8ef9d15' }), index.h("div", { key: '9b10ea4a41c5bd4d0c66effdd0d4ca308ad03706' }, index.h("alc-icon", { key: '760138d17d0de1b12bf0757e7de19d99cc903059', name: "box-arrow-right", label: "" }), index.h("a", { key: '9733df1f115340fe20a43165ae8ab935d6ca91a1', class: "alc-link ml-2", href: this.logoutUrl, onClick: (event) => this.handleLogout(event) }, "Sair")))));
    }
};
AlcUser.style = alcUserCss;

exports.alc_user = AlcUser;
