// alcance/dist/cjs/alc-scroll-panel_2.cjs.entry.js
'use strict';

var index = require('./index-BOXTLnv4.js');
var logger = require('./logger-D5AeZhi-.js');
var throttle = require('./throttle-DAOUKsy2.js');

const alcScrollPanelBaseCss = "alc-scroll-panel{display:block}.alc-scroll-panel{display:-ms-flexbox;display:flex;-ms-flex-direction:row;flex-direction:row;-ms-flex-align:stretch;align-items:stretch}.alc-scroll-panel__content{overflow-x:hidden;white-space:nowrap}";

const themeAlcScrollPanelCss = "alc-scroll-panel{display:block}.alc-scroll-panel{display:-ms-flexbox;display:flex;-ms-flex-direction:row;flex-direction:row;-ms-flex-align:stretch;align-items:stretch}.alc-scroll-panel__content{overflow-x:hidden;white-space:nowrap}[data-alc-theme=\"dark\"] alc-scroll-panel{--border-color:var(--alc-color-border);--shadow-color:#adadad;--outline-color:var(--alc-color-action-30)}[data-alc-theme=\"dark\"] .alc-scroll-panel__button{--text-color:var(--alc-color-link-text);--bg-color:transparent}[data-alc-theme=\"dark\"] .alc-scroll-panel__button:hover{--text-color:var(--alc-color-text-0);--bg-color:var(--alc-color-action-50)}[data-alc-theme=\"dark\"] .alc-scroll-panel__button:active{--text-color:var(--alc-color-text-0);--bg-color:var(--alc-color-action-70)}[data-alc-theme=\"dark\"] .alc-scroll-panel__button:disabled{--text-color:var(--alc-color-neutral-60);--bg-color:transparent}[data-alc-theme=\"light\"] alc-scroll-panel{--border-color:var(--alc-color-border);--shadow-color:#adadad;--outline-color:var(--alc-color-action-60)}[data-alc-theme=\"light\"] .alc-scroll-panel__button{--text-color:var(--alc-color-link-text);--bg-color:transparent}[data-alc-theme=\"light\"] .alc-scroll-panel__button:hover{--text-color:var(--alc-color-text-contrast);--bg-color:var(--alc-color-action-50)}[data-alc-theme=\"light\"] .alc-scroll-panel__button:active{--text-color:var(--alc-color-text-contrast);--bg-color:var(--alc-color-action-70)}[data-alc-theme=\"light\"] .alc-scroll-panel__button:disabled{--text-color:var(--alc-color-neutral-30);--bg-color:transparent}.alc-scroll-panel__button{--border-width:var(--alc-border-width);border-width:var(--border-width) 0;border-bottom-color:var(--border-color);border-top-color:transparent;height:100%;width:2.5rem;background-color:var(--bg-color);color:var(--text-color);transition-property:color, background-color, border-color, fill, stroke, -webkit-text-decoration-color;-webkit-transition-property:color, background-color, border-color, fill, stroke, -webkit-text-decoration-color;transition-property:color, background-color, border-color, text-decoration-color, fill, stroke;transition-property:color, background-color, border-color, text-decoration-color, fill, stroke, -webkit-text-decoration-color;-webkit-transition-timing-function:cubic-bezier(0.4, 0, 0.2, 1);transition-timing-function:cubic-bezier(0.4, 0, 0.2, 1);-webkit-transition-duration:150ms;transition-duration:150ms}.alc-scroll-panel__wrapper-content{position:relative;overflow:hidden}.alc-scroll-panel__content{overflow:auto}.alc-scroll-panel__content::-webkit-scrollbar{display:none}.alc-scroll-panel__has-scroll-left::before,.alc-scroll-panel__has-scroll-right::after{display:block;--tw-content:\"\\200C\";content:var(--tw-content);pointer-events:none;position:absolute;top:0px;z-index:10;height:100%;width:0.75rem}.alc-scroll-panel__has-scroll-left::before{left:0px;-webkit-box-shadow:inset 8px 0px 8px -8px var(--shadow-color);box-shadow:inset 8px 0px 8px -8px var(--shadow-color)}.alc-scroll-panel__has-scroll-right::after{right:0px;-webkit-box-shadow:inset -8px 0px 8px -8px var(--shadow-color);box-shadow:inset -8px 0px 8px -8px var(--shadow-color)}.alc-scroll-panel__button:focus-visible,.alc-scroll-panel__button.active:focus-visible{outline-style:solid;outline-width:2px;outline-offset:-2px;outline-color:var(--outline-color)}";

const SCROLL_STEP = 20;
const AlcScrollPanel = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.hasScroll = false;
        this.canScrollRight = false;
        this.canScrollLeft = false;
        /**
        * Define se há foco.
        */
        this.hasFocus = true;
        this.scrollWidth = 0;
        this.scrolling = 0;
        this.resizeThrottle = throttle.throttle(this.resize, 200);
        this.scrollThrottle = throttle.throttle(this.handleScroll, 300);
    }
    themeLoadedHandler() {
        this.handleScroll();
    }
    handleResize() {
        this.resizeThrottle();
    }
    handleScrollToElement(element) {
        if (!!element) {
            element.scrollIntoView({ inline: "center", block: 'nearest' });
        }
    }
    resize() {
        this.handleScroll();
        this.handleScrollToElement(this.scrollToElement);
    }
    handleScroll() {
        this.scrollWidth = this.content.scrollWidth;
        this.hasScroll = this.scrollWidth > this.content.offsetWidth;
        this.setCanScroll();
    }
    scrollRight() {
        this.content.scrollLeft += SCROLL_STEP;
        this.setCanScroll();
    }
    scrollLeft() {
        this.content.scrollLeft = this.content.scrollLeft - SCROLL_STEP;
        this.setCanScroll();
    }
    startScrolling(button, scroll) {
        if (this.scrolling)
            return;
        this.scrolling = window.setInterval(() => {
            if (button.disabled) {
                this.stopScrolling();
                return;
            }
            scroll.apply(this);
        }, 200);
    }
    startScrollingRight(e) {
        const button = e.currentTarget;
        if (button instanceof HTMLButtonElement) {
            this.startScrolling(button, this.scrollRight);
        }
    }
    startScrollingLeft(e) {
        const button = e.currentTarget;
        if (button instanceof HTMLButtonElement) {
            this.startScrolling(button, this.scrollLeft);
        }
    }
    stopScrolling() {
        window.clearInterval(this.scrolling);
        this.scrolling = 0;
    }
    setCanScroll() {
        // Math.ceil foi necessário ao verificar funcionamento em um dispositivo móvel real,
        // onde scrollLeft apresenta números fracionários.
        this.canScrollRight = Math.ceil(this.content.offsetWidth + this.content.scrollLeft) < this.scrollWidth;
        this.canScrollLeft = this.content.scrollLeft > 0;
    }
    componentDidRender() {
        var _a;
        /*
         O trecho abaixo permite que o componente seja atualizado
         sempre que houver mudanças no conteúdo (como a inclusão de "fihos").
         Isso foi feito para permitir, por exemplo, que o conteúdo do panel
         seja atualizado dinamicamente com a manipulação do DOM.
         */
        (_a = this.mo) === null || _a === void 0 ? void 0 : _a.disconnect();
        this.mo = new MutationObserver(() => {
            logger.logger.debug('forceUpdate scroll-panel');
            index.forceUpdate(this.el);
        });
        this.mo.observe(this.el, {
            childList: true,
            subtree: true
        });
        this.handleScroll();
    }
    disconnectedCallback() {
        var _a;
        (_a = this.mo) === null || _a === void 0 ? void 0 : _a.disconnect();
        this.resizeThrottle.cancel();
        this.scrollThrottle.cancel();
    }
    render() {
        return (index.h(index.Host, { key: '3c5dee05949d2b073dcb87abd7c4df0826377f4c' }, index.h("div", { key: '8cb56b2dfa1b2b82f8b3eb7dd52018410e1a9554', class: "alc-scroll-panel" }, this.hasScroll
            ? index.h("div", { key: "left-button" }, index.h("button", { onClick: this.scrollLeft.bind(this), onMouseDown: this.startScrollingLeft.bind(this), onMouseUp: this.stopScrolling.bind(this), onTouchStart: this.startScrollingLeft.bind(this), onTouchEnd: this.stopScrolling.bind(this), onTouchCancel: this.stopScrolling.bind(this), class: "alc-scroll-panel__button", disabled: !this.canScrollLeft, "aria-label": "Ir para esquerda", tabindex: this.hasFocus ? '0' : '-1' }, index.h("alc-icon", { name: "chevron-left", label: '' })))
            : null, index.h("div", { class: {
                "alc-scroll-panel__wrapper-content": true,
                "alc-scroll-panel__has-scroll-left": this.canScrollLeft,
                "alc-scroll-panel__has-scroll-right": this.canScrollRight
            }, key: "content" }, index.h("div", { key: '93e54d7ba7f47158fc3fcdfd357a742ed83568cc', class: "alc-scroll-panel__content", ref: el => this.content = el, onScroll: () => this.scrollThrottle() }, index.h("slot", { key: '49672486ccf36230946da3949ee80fc61f4098c4' }))), this.hasScroll
            ? index.h("div", { key: "right-button" }, index.h("button", { onClick: this.scrollRight.bind(this), onMouseDown: this.startScrollingRight.bind(this), onMouseUp: this.stopScrolling.bind(this), onTouchStart: this.startScrollingRight.bind(this), onTouchEnd: this.stopScrolling.bind(this), onTouchCancel: this.stopScrolling.bind(this), class: "alc-scroll-panel__button", disabled: !this.canScrollRight, "aria-label": "Ir para direita", tabindex: this.hasFocus ? '0' : '-1' }, index.h("alc-icon", { name: "chevron-right", label: '' })))
            : null)));
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "scrollToElement": ["handleScrollToElement"]
    }; }
};
AlcScrollPanel.style = {
    base: alcScrollPanelBaseCss,
    theme: themeAlcScrollPanelCss
};

const alcTabButtonBaseCss = "alc-tab-button{display:inline-block}.alc-tabs__button.active{-webkit-text-decoration-line:underline;text-decoration-line:underline}";

const themeAlcTabButtonCss = "[data-alc-theme=\"dark\"] .alc-tabs__button{--text-color:#52daf2;--bg-color-button:var(--alc-color-neutral-70)}[data-alc-theme=\"dark\"] .alc-tabs__button:hover{--bg-color-button:#3d4551}[data-alc-theme=\"dark\"] .alc-tabs__button.active{--text-color:var(--alc-color-text-primary);--bg-color-button:var(--bg-color)}[data-alc-theme=\"light\"] .alc-tabs__button{--text-color:var(--alc-color-link-text);--bg-color-button:#dfe1e2}[data-alc-theme=\"light\"] .alc-tabs__button:hover{--bg-color-button:#c6cace}[data-alc-theme=\"light\"] .alc-tabs__button.active{--text-color:var(--alc-color-text-primary);--bg-color-button:var(--bg-color)}alc-tab-button{display:inline-block}.alc-tabs__button{z-index:10;border-color:var(--border-color);background-color:var(--bg-color-button);padding:0.5rem;padding-top:0.75rem;color:var(--text-color);border-right-width:var(--border-width);border-bottom-width:var(--border-width)}.alc-tabs__button.active{cursor:default;padding-top:0.5rem;-webkit-text-decoration-line:none;text-decoration-line:none;border-top-width:4px;border-bottom-color:transparent}alc-tab-button:first-child{border-color:var(--border-color);border-left-width:var(--border-width)}.alc-tabs__button:focus-visible,.alc-tabs__button.active:focus-visible{outline-style:solid;outline-width:2px;outline-offset:-8px;outline-color:var(--outline-color)}";

const AlcTabButton = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.alcClick = index.createEvent(this, "alc-click", 7);
        this.alcNext = index.createEvent(this, "alc-next", 7);
        this.alcPrevious = index.createEvent(this, "alc-previous", 7);
        this.alcFirst = index.createEvent(this, "alc-first", 7);
        this.alcLast = index.createEvent(this, "alc-last", 7);
    }
    selectedChanged(newValue, oldValue) {
        if (newValue !== oldValue) {
            if (newValue) {
                this.button.setAttribute('tabindex', '0');
            }
            else {
                this.button.setAttribute('tabindex', '-1');
            }
        }
    }
    selectTab(e) {
        if (this.tab !== undefined) {
            this.alcClick.emit({
                tab: this.tab
            });
            e.preventDefault();
        }
    }
    // Inspirado na implementação do WAI - função onKeyDown().
    // https://www.w3.org/WAI/content-assets/wai-aria-practices/patterns/tabs/examples/js/tabs-automatic.js
    keyDownHandler(e) {
        let handled = false;
        switch (e.key) {
            case 'ArrowRight':
                this.alcNext.emit();
                handled = true;
                break;
            case 'ArrowLeft':
                this.alcPrevious.emit();
                handled = true;
                break;
            case 'Home':
                this.alcFirst.emit();
                handled = true;
                break;
            case 'End':
                this.alcLast.emit();
                handled = true;
                break;
        }
        if (handled) {
            e.preventDefault();
        }
    }
    componentWillLoad() {
    }
    render() {
        var _a;
        (_a = this.tab) !== null && _a !== void 0 ? _a : logger.logger.report('tab', this.el.tagName.toLowerCase(), this.el);
        const getId = () => {
            let id;
            if (this.el.id) {
                id = this.el.id;
            }
            else {
                id = `alc-button_${this.el.tab}`;
            }
            return id;
        };
        return (index.h(index.Host, null, index.h("button", { onClick: this.selectTab.bind(this), onKeyDown: this.keyDownHandler.bind(this), role: 'tab', "aria-selected": this.selected ? 'true' : 'false', tabindex: '-1', id: getId(), class: {
                'alc-tabs__button': true,
                active: this.selected
            }, ref: el => this.button = el }, index.h("slot", null))));
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "selected": ["selectedChanged"]
    }; }
};
AlcTabButton.style = {
    base: alcTabButtonBaseCss,
    theme: themeAlcTabButtonCss
};

exports.alc_scroll_panel = AlcScrollPanel;
exports.alc_tab_button = AlcTabButton;
