// alcance/dist/cjs/logger-D5AeZhi-.js
'use strict';

require('./index-BOXTLnv4.js');

// @ts-nocheck
const PREFIX = [
    '%cAlcance',
    'color: white;background:#40807e; font-size:10px; padding:2px 4px; border-radius: 2px',
];
const noop = () => { };
const forceDebug = (localStorage === null || localStorage === void 0 ? void 0 : localStorage.getItem('alc-debug')) === 'true';
// log somente vai pra console se estiver em desenvolvimento
const logFn = () => {
    if (forceDebug) {
        return console.log.bind(undefined, ...PREFIX);
    }
    return noop;
};
// debug somente vai pra console se estiver em desenvolvimento
const debugFn = () => {
    if (forceDebug) {
        return console.debug.bind(undefined, ...PREFIX);
    }
    return noop;
};
var logger = {
    log: logFn(),
    debug: debugFn(),
    warn: console.warn.bind(undefined, ...PREFIX),
    error: console.error.bind(undefined, ...PREFIX),
    report: console.error.bind(undefined, ...[
        `${PREFIX[0]}%c A propriedade obrigatória "%s" não foi informada. Componente: %s`, // Mensagem completa de log
        PREFIX[1], // Estilo para o prefixo
        'color:unset; background:unset; font-size:unset; padding:unset; border-radius:unset' // Estilo para a mensagem
    ]),
};

exports.logger = logger;
