// alcance/dist/collection/components/alc-header-v1/alc-header-v1.js
import { getAssetPath, Host, h } from "@stencil/core";
import test from "../utils/testAttributes";
export class AlcHeaderV1 {
    async handleHome(event) {
        const triggerEvent = this.alcHome.emit();
        if (triggerEvent.defaultPrevented) {
            event.preventDefault();
        }
    }
    renderLink() {
        if (this.homeUrl !== undefined) {
            return (h("a", Object.assign({ href: this.homeUrl, onClick: event => this.handleHome(event) }, test('data-test-link')), this.name));
        }
        return this.name;
    }
    render() {
        return (h(Host, { key: '6524921c0282f57a3b1b703b5bad91fb18bf5dda', class: "alc-header-v1", "data-alc-top": true }, h("header", { key: '477f40e2d7bebc0a3788172e22d100857f6f39e2' }, h("slot", { key: '8743e7221f04fe764f6afd729e37c90b34ac3886' }), h("div", { key: '36c6af926a41f29327abafdc9e2428329100276f', class: "alc-header-v1__container" }, h("img", { key: 'b28a882d9f407a343bc81365a46207504fedbe46', src: getAssetPath('assets/images/logo_camara_peq.png'), alt: "C\u00E2mara dos Deputados", class: "alc-header-v1__logo" }), h("div", { key: '332cc1e4442682d56b9ab6252f95372925ca7bb8', class: "alc-header-v1__logo-full" }, h("img", { key: 'fae90a03686f65678dac91e7429ffa1d52b643e7', src: getAssetPath('assets/images/logo_camara.png'), alt: "C\u00E2mara dos Deputados" })), h("span", Object.assign({ key: 'a28701aab859fd2d680f4f300c9f04d156f36aaa', class: "alc-header-v1__title" }, test('data-test-title')), this.name !== undefined ? (this.renderLink()) : (h("slot", { name: "link" })))))));
    }
    static get is() { return "alc-header-v1"; }
    static get originalStyleUrls() {
        return {
            "$": ["alc-header-v1.css"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["alc-header-v1.css"]
        };
    }
    static get assetsDirs() { return ["assets/images"]; }
    static get properties() {
        return {
            "name": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "O nome do sistema. Ser\u00E1 mostrado no cabe\u00E7alho, que deve estar presente em todas do sistema."
                },
                "getter": false,
                "setter": false,
                "attribute": "name",
                "reflect": true
            },
            "homeUrl": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "URL para a p\u00E1gina inicial do sistema. O nome do sistema ser\u00E1 transformado em um link para essa URL."
                },
                "getter": false,
                "setter": false,
                "attribute": "home-url",
                "reflect": true
            }
        };
    }
    static get events() {
        return [{
                "method": "alcHome",
                "name": "alc-home",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Evento disparado quando o usu\u00E1rio aciona o link para a p\u00E1gina inicial.\nExecutar `preventDefault()` evita que a navega\u00E7\u00E3o para a p\u00E1gina inicial aconte\u00E7a."
                },
                "complexType": {
                    "original": "null",
                    "resolved": "null",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "el"; }
}
