// alcance/dist/collection/components/alc-environment-label/alc-environment-label.js
import { Host, h } from "@stencil/core";
import { Environments } from "./environments";
export class AlcEnvironmentLabel {
    constructor() {
        this.environments = {
            [Environments.Prototype]: 'Protótipo',
            [Environments.Development]: 'Desenvolvimento',
            [Environments.Testing]: 'Teste',
            [Environments.Homologation]: 'Homologação',
            [Environments.Production]: '', // É válido (reconhecido), mas não deve ter texto apresentado.
        };
    }
    render() {
        var _a;
        return (h(Host, { key: '978bf8a1bdb0658e3362d2dbcede1486c73ca434', class: {
                'alc-environment-label--prototype': this.env === Environments.Prototype,
                'alc-environment-label--development': this.env === Environments.Development,
                'alc-environment-label--testing': this.env === Environments.Testing,
                'alc-environment-label--homologation': this.env === Environments.Homologation,
                'alc-environment-label--production': this.env === Environments.Production,
            } }, (_a = this.environments[this.env]) !== null && _a !== void 0 ? _a : 'Não reconhecido'));
    }
    static get is() { return "alc-environment-label"; }
    static get originalStyleUrls() {
        return {
            "$": ["alc-environment-label.css"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["alc-environment-label.css"]
        };
    }
    static get properties() {
        return {
            "env": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "EnvType",
                    "resolved": "\"development\" | \"homologation\" | \"production\" | \"prototype\" | \"testing\"",
                    "references": {
                        "EnvType": {
                            "location": "import",
                            "path": "./environments",
                            "id": "src/components/alc-environment-label/environments.ts::EnvType"
                        }
                    }
                },
                "required": true,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "getter": false,
                "setter": false,
                "attribute": "env",
                "reflect": true
            }
        };
    }
}
