// alcance/dist/collection/components/alc-sort-indicator/alc-sort-indicator.js
import { Host, h } from "@stencil/core";
const iconMap = new Map();
iconMap.set('asc', {
    name: 'chevron-down'
});
iconMap.set('desc', {
    name: 'chevron-up'
});
iconMap.set('none', {
    name: 'chevron-expand'
});
export class AlcSortIndicator {
    constructor() {
        /**
        * Define indicação do sort.
        */
        this.sorting = 'none';
    }
    render() {
        return (h(Host, { key: 'd3de3f859d218235a32d3cbd2b1c78a1a667c903' }, h("span", { key: 'b08d705fef11a025faa69144ab539dadc8e22035', class: "alc-sort-indicator" }, h("alc-icon", { key: 'd9f35015d02ef4f5f8b82cb0a1ef1245e0fc9354', name: iconMap.get(this.sorting).name, class: "alc-sort-indicator__icon", label: "" }))));
    }
    static get is() { return "alc-sort-indicator"; }
    static get properties() {
        return {
            "sorting": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'asc' | 'desc' | 'none'",
                    "resolved": "\"asc\" | \"desc\" | \"none\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Define indica\u00E7\u00E3o do sort."
                },
                "getter": false,
                "setter": false,
                "attribute": "sorting",
                "reflect": true,
                "defaultValue": "'none'"
            }
        };
    }
}
