// alcance/dist/collection/components/alc-view-more/alc-view-more.js
import { Host, h } from "@stencil/core";
import { getUniqueId } from "../utils/getUniqueId";
export class AlcViewMore {
    constructor() {
        /**
         * Nome da ação para revelar o conteúdo.
         */
        this.toggleTextOpen = 'Veja mais...';
        /**
         * Nome da ação para esconder o conteúdo.
         */
        this.toggleTextClose = 'Fechar';
        /**
         * Alinhamento do botão de ação.
         */
        this.togglePosition = 'center';
        /**
         * Classes CSS para serem aplicadas ao texto da ação.
         */
        this.textClass = '';
        /**
         * Tamanho mínimo do conteúdo a ser exibido.
         * Pode ser em px, rem ou em: 200px; 20rem; 10em;
         */
        this.minHeight = '8em';
        /**
         * Nome da variável CSS correspondente à cor de fundo em que o componente está inserido.
         * Aplica-se ao degradê do corte do texto.
         */
        this.bgColorVar = '--alc-color-surface-primary';
        this.contentId = getUniqueId();
    }
    toggle(e) {
        e.preventDefault();
        this.contentExpanded = !this.contentExpanded;
    }
    setPosition() {
        if (this.togglePosition === 'right') {
            return 'justify-end';
        }
        if (this.togglePosition === 'left') {
            return 'justify-start';
        }
        return 'justify-center';
    }
    getContentCssClasses() {
        let classes = `alc-view-more__content ${this.textClass}`;
        if (this.contentExpanded) {
            classes = `${classes} alc-view-more__content--opened`;
        }
        return classes;
    }
    componentWillLoad() {
        this.contentExpanded = false;
    }
    componentDidRender() {
        let content = this.el.querySelector('.alc-view-more__content');
        let maxH = content.scrollHeight;
        content.style.setProperty('--min-height', this.minHeight);
        content.style.setProperty('--max-height', maxH + 'px');
        content.style.setProperty('--bg-color', `var(${this.bgColorVar})`);
    }
    render() {
        return (h(Host, { key: '6ed1099e4c471134cf475145ba8a7df40866e139' }, h("div", { key: 'cdcabe99fd1fa7185cc590e41eabdc68d5329d09', id: this.contentId, class: this.getContentCssClasses() }, h("slot", { key: '784a5200aadfa30cf70365c147129b97fe48edea' })), h("div", { key: 'f67e9efef0ffefafa163244bf01e4a65cefd996d', class: 'alc-view-more__toggle ' + (this.setPosition()) }, h("a", { key: '14f59528e78e47e0093ad877c9e436c5e2c9dfb1', href: '#' + this.contentId, onClick: e => { this.toggle(e); }, "aria-expanded": this.contentExpanded.toString(), "aria-controls": this.contentId, role: "button" }, this.contentExpanded ? this.toggleTextClose : this.toggleTextOpen))));
    }
    static get is() { return "alc-view-more"; }
    static get originalStyleUrls() {
        return {
            "$": ["alc-view-more.css"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["alc-view-more.css"]
        };
    }
    static get properties() {
        return {
            "toggleTextOpen": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Nome da a\u00E7\u00E3o para revelar o conte\u00FAdo."
                },
                "getter": false,
                "setter": false,
                "attribute": "toggle-text-open",
                "reflect": true,
                "defaultValue": "'Veja mais...'"
            },
            "toggleTextClose": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Nome da a\u00E7\u00E3o para esconder o conte\u00FAdo."
                },
                "getter": false,
                "setter": false,
                "attribute": "toggle-text-close",
                "reflect": true,
                "defaultValue": "'Fechar'"
            },
            "togglePosition": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Alinhamento do bot\u00E3o de a\u00E7\u00E3o."
                },
                "getter": false,
                "setter": false,
                "attribute": "toggle-position",
                "reflect": true,
                "defaultValue": "'center'"
            },
            "textClass": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Classes CSS para serem aplicadas ao texto da a\u00E7\u00E3o."
                },
                "getter": false,
                "setter": false,
                "attribute": "text-class",
                "reflect": true,
                "defaultValue": "''"
            },
            "minHeight": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Tamanho m\u00EDnimo do conte\u00FAdo a ser exibido.\nPode ser em px, rem ou em: 200px; 20rem; 10em;"
                },
                "getter": false,
                "setter": false,
                "attribute": "min-height",
                "reflect": true,
                "defaultValue": "'8em'"
            },
            "bgColorVar": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Nome da vari\u00E1vel CSS correspondente \u00E0 cor de fundo em que o componente est\u00E1 inserido.\nAplica-se ao degrad\u00EA do corte do texto."
                },
                "getter": false,
                "setter": false,
                "attribute": "bg-color-var",
                "reflect": true,
                "defaultValue": "'--alc-color-surface-primary'"
            }
        };
    }
    static get states() {
        return {
            "contentExpanded": {}
        };
    }
    static get elementRef() { return "el"; }
}
