// alcance/dist/collection/components/alc-tab/stories/alc-tab.stories.js
// @ts-ignore
import { argTypes } from "./alc-tab.args";
import { render as renderArgs } from "../../../stories/functions/components.renderArgs";
import getDefaultArgs from "../../../stories/functions/components.defaultArgs";
export default {
    title: 'Componentes Alfa/Tab',
    name: 'Tab',
    component: 'alc-tab',
    tags: ['alfa'],
    argTypes: argTypes,
};
const defaultArgs = getDefaultArgs(argTypes);
// ESTRUTURA BÁSICA PARA CRIAR EXEMPLOS DO COMPONENTE
// PARA NOVOS EXEMPLOS, REPLIQUE O CÓDIGO ABAIXO E SUBSTITUA O 'Basico' POR OUTRO TIPO DE EXEMPLO: 'Avancado'
// Exemplo Básico
export const Basico = {
    name: 'Básico',
    args: Object.assign(Object.assign({}, defaultArgs), { label: 'Aba 1', tab: 'tab-1', selected: true, 'content-focus': false }),
    render: args => `
<alc-tabs>
  <alc-tab-button slot="button" tab="tab-1">Aba 1</alc-tab-button>
  <alc-tab ${renderArgs(argTypes, args)}>
    <p>Conteúdo da Aba 1</p>
  </alc-tab>
</alc-tabs>
  `,
};
// Exemplo com Múltiplas Abas
export const MultiplasAbas = {
    name: 'Múltiplas Abas',
    args: Object.assign(Object.assign({}, defaultArgs), { label: 'Aba 1', tab: 'tab-1', selected: false, 'content-focus': false }),
    render: args => `
<alc-tabs>
  <alc-tab-button slot="button" tab="tab-1">Aba 1</alc-tab-button>
  <alc-tab-button slot="button" tab="tab-2">Aba 2</alc-tab-button>
  <alc-tab-button slot="button" tab="tab-3">Aba 3</alc-tab-button>
  <alc-tab-button slot="button" tab="tab-4">Aba 4</alc-tab-button>
  <alc-tab ${renderArgs(argTypes, args)}>
    <p>Conteúdo da Aba 1</p>
  </alc-tab>
  <alc-tab label="Aba 2" tab="tab-2">
    <p>Conteúdo da Aba 2</p>
  </alc-tab>
  <alc-tab label="Aba 3" tab="tab-3">
    <p>Conteúdo da Aba 3</p>
  </alc-tab>
  <alc-tab label="Aba 4" tab="tab-4">
    <p>Conteúdo da Aba 4</p>
  </alc-tab>
</alc-tabs>
  `,
};
// Exemplo com Foco no Conteúdo
export const ComFoco = {
    name: 'Com Foco',
    args: Object.assign(Object.assign({}, defaultArgs), { label: 'Aba 1', tab: 'tab-1', selected: true, 'content-focus': true }),
    render: args => `
<alc-tabs>
  <alc-tab-button slot="button" tab="tab-1">Aba 1</alc-tab-button>
  <alc-tab-button slot="button" tab="tab-2">Aba 2</alc-tab-button>
  <alc-tab ${renderArgs(argTypes, args)}>
    <p>Conteúdo da Aba 1</p>
    <a class="alc-link" href="#">item focalizável</a>
  </alc-tab>
  <alc-tab label="Aba 2" tab="tab-2">
    <p>Conteúdo da Aba 2</p>
  </alc-tab>
</alc-tabs>
  `,
};
