// alcance-vue-library/lib/components.js
import { defineContainer } from './vue-component-lib/utils';
export const AlcAlert = defineContainer('alc-alert', undefined, [
    'visible',
    'type',
    'dismissible',
    'alc-hide',
    'alc-after-hide'
]);
export const AlcAutocomplete = defineContainer('alc-autocomplete', undefined, [
    'items',
    'displayKeys',
    'label',
    'hint',
    'required',
    'placeholder',
    'listDirection',
    'disabled',
    'loading',
    'error',
    'errorMsg',
    'alc-change'
]);
export const AlcBadge = defineContainer('alc-badge', undefined, [
    'label',
    'count',
    'color',
    'dot',
    'hidden',
    'outlined',
    'position',
    'pulsate'
]);
export const AlcBreadcrumb = defineContainer('alc-breadcrumb', undefined);
export const AlcBreadcrumbItem = defineContainer('alc-breadcrumb-item', undefined, [
    'url',
    'label',
    'iconName'
]);
export const AlcCheckbox = defineContainer('alc-checkbox', undefined, [
    'hint',
    'label',
    'errorMsg'
]);
export const AlcDatatable = defineContainer('alc-datatable', undefined, [
    'data',
    'options',
    'deferStart',
    'tableId',
    'alc-child-row',
    'alc-column-sizing',
    'alc-column-visibility',
    'alc-destroy',
    'alc-draw',
    'alc-error',
    'alc-init',
    'alc-length',
    'alc-order',
    'alc-page',
    'alc-pre-draw',
    'alc-pre-init',
    'alc-pre-xhr',
    'alc-processing',
    'alc-request-child',
    'alc-search',
    'alc-state-load-params',
    'alc-state-loaded',
    'alc-state-save-params',
    'alc-xhr',
    'alc-deselect',
    'alc-select',
    'alc-select-blur',
    'alc-select-items',
    'alc-select-style',
    'alc-user-select'
]);
export const AlcDetails = defineContainer('alc-details', undefined, [
    'summary',
    'disabled',
    'opened',
    'alc-show',
    'alc-close'
]);
export const AlcDrawer = defineContainer('alc-drawer', undefined, [
    'elementToFocus',
    'isVisible',
    'alc-show',
    'alc-after-show',
    'alc-hide',
    'alc-after-hide'
]);
export const AlcDropdown = defineContainer('alc-dropdown', undefined, [
    'open',
    'hideOn',
    'alc-show',
    'alc-after-show',
    'alc-hide',
    'alc-after-hide'
]);
export const AlcEnvironmentLabel = defineContainer('alc-environment-label', undefined, [
    'env'
]);
export const AlcEventLab = defineContainer('alc-event-lab', undefined, [
    'alc-change',
    'alc-after-change'
]);
export const AlcExpander = defineContainer('alc-expander', undefined, [
    'label',
    'open',
    'hideLabel',
    'alc-show',
    'alc-after-show',
    'alc-hide',
    'alc-after-hide'
]);
export const AlcField = defineContainer('alc-field', undefined, [
    'label',
    'hint',
    'required',
    'errorMsg'
]);
export const AlcFieldChecker = defineContainer('alc-field-checker', undefined, [
    'checkOnSubmit'
]);
export const AlcFieldset = defineContainer('alc-fieldset', undefined, [
    'legend',
    'hint',
    'required',
    'errorMsg'
]);
export const AlcHeader = defineContainer('alc-header', undefined, [
    'name',
    'description',
    'homeUrl',
    'alc-home'
]);
export const AlcHeaderAction = defineContainer('alc-header-action', undefined, [
    'iconName',
    'variant',
    'url',
    'alc-select'
]);
export const AlcHeaderId = defineContainer('alc-header-id', undefined, [
    'name',
    'description',
    'homeUrl',
    'alc-home'
]);
export const AlcHeaderNotifications = defineContainer('alc-header-notifications', undefined, [
    'notifications',
    'variant',
    'url',
    'alc-click'
]);
export const AlcHeaderV1 = defineContainer('alc-header-v1', undefined, [
    'name',
    'homeUrl',
    'alc-home'
]);
export const AlcIcon = defineContainer('alc-icon', undefined, [
    'label',
    'flipRtl',
    'name',
    'src',
    'icon',
    'lazy',
    'sanitize'
]);
export const AlcInputFile = defineContainer('alc-input-file', undefined, [
    'multiple',
    'idInput',
    'accept',
    'disabled',
    'required',
    'buttonType',
    'iconName',
    'mode',
    'alc-file-selected'
]);
export const AlcLoading = defineContainer('alc-loading', undefined, [
    'active',
    'label',
    'endMsg',
    'variant'
]);
export const AlcMenu = defineContainer('alc-menu', undefined);
export const AlcMenuItem = defineContainer('alc-menu-item', undefined, [
    'type',
    'checked',
    'disabled',
    'value',
    'alc-select'
]);
export const AlcMenuItemTheme = defineContainer('alc-menu-item-theme', undefined);
export const AlcMenuLink = defineContainer('alc-menu-link', undefined, [
    'disabled',
    'value',
    'alc-select'
]);
export const AlcModal = defineContainer('alc-modal', undefined, [
    'open',
    'headerText',
    'size',
    'preventOverlayClose',
    'alc-after-show',
    'alc-hide',
    'alc-after-hide',
    'alc-focus-after-hide'
]);
export const AlcNav = defineContainer('alc-nav', undefined);
export const AlcNavPanel = defineContainer('alc-nav-panel', undefined, [
    'open',
    'alc-state-change',
    'alc-state-request'
]);
export const AlcPagination = defineContainer('alc-pagination', undefined, [
    'totalPages',
    'currentPage',
    'alc-change'
]);
export const AlcPopup = defineContainer('alc-popup', undefined, [
    'active',
    'placement',
    'flip',
    'shift',
    'strategy',
    'distance',
    'sync',
    'arrow'
]);
export const AlcRadio = defineContainer('alc-radio', undefined, [
    'hint',
    'label',
    'errorMsg'
]);
export const AlcScrollPanel = defineContainer('alc-scroll-panel', undefined, [
    'hasFocus',
    'scrollToElement'
]);
export const AlcSidepanel = defineContainer('alc-sidepanel', undefined, [
    'visible',
    'alc-state-change',
    'alc-state-request'
]);
export const AlcSkipLink = defineContainer('alc-skip-link', undefined, [
    'anchor'
]);
export const AlcSkipToNav = defineContainer('alc-skip-to-nav', undefined);
export const AlcSortIndicator = defineContainer('alc-sort-indicator', undefined, [
    'sorting'
]);
export const AlcTab = defineContainer('alc-tab', undefined, [
    'label',
    'tab',
    'selected',
    'contentFocus'
]);
export const AlcTabButton = defineContainer('alc-tab-button', undefined, [
    'tab',
    'selected',
    'alc-click',
    'alc-next',
    'alc-previous',
    'alc-first',
    'alc-last'
]);
export const AlcTable = defineContainer('alc-table', undefined);
export const AlcTabs = defineContainer('alc-tabs', undefined, [
    'selected',
    'alc-change'
]);
export const AlcTheme = defineContainer('alc-theme', undefined);
export const AlcTooltip = defineContainer('alc-tooltip', undefined, [
    'active',
    'content',
    'placement',
    'trigger',
    'strategy',
    'alc-show',
    'alc-after-show',
    'alc-hide',
    'alc-after-hide'
]);
export const AlcUser = defineContainer('alc-user', undefined, [
    'name',
    'registrationNumber',
    'logoutUrl',
    'imgSrc',
    'alc-logout'
]);
export const AlcUserMenu = defineContainer('alc-user-menu', undefined, [
    'variation',
    'name',
    'registrationNumber',
    'logoutUrl',
    'imgSrc',
    'alc-logout'
]);
export const AlcViewMore = defineContainer('alc-view-more', undefined, [
    'toggleTextOpen',
    'toggleTextClose',
    'togglePosition',
    'textClass',
    'minHeight',
    'bgColorVar'
]);
