// alcance/dist/collection/components/alc-sort-indicator/stories/alc-sort-indicator.stories.js
// @ts-ignore
import { argTypes } from "./alc-sort-indicator.args";
import { render as renderArgs } from "../../../stories/functions/components.renderArgs";
import getDefaultArgs from "../../../stories/functions/components.defaultArgs";
export default {
    title: 'Componentes Alfa/Sort Indicator',
    name: 'Sort Indicator',
    component: 'alc-sort-indicator',
    tags: ['alfa'],
    argTypes: argTypes,
};
const defaultArgs = getDefaultArgs(argTypes);
// Exemplo Básico
export const Basico = {
    name: 'Básico',
    args: Object.assign(Object.assign({}, defaultArgs), { sorting: 'none' }),
    render: args => `
    <alc-sort-indicator ${renderArgs(argTypes, args)}></alc-sort-indicator>
  `,
};
// Exemplo Ascendente
export const Ascendente = {
    name: 'Ascendente',
    args: Object.assign(Object.assign({}, defaultArgs), { sorting: 'asc' }),
    render: args => `
    <alc-sort-indicator ${renderArgs(argTypes, args)}></alc-sort-indicator>
  `,
};
// Exemplo Descendente
export const Descendente = {
    name: 'Descendente',
    args: Object.assign(Object.assign({}, defaultArgs), { sorting: 'desc' }),
    render: args => `
    <alc-sort-indicator ${renderArgs(argTypes, args)}></alc-sort-indicator>
  `,
};
