// alcance/dist/collection/components/alc-header-id/alc-header-id.js
import { Host, h, getAssetPath } from "@stencil/core";
export class AlcHeaderId {
    async handleHome(event) {
        const triggerEvent = this.alcHome.emit();
        if (triggerEvent.defaultPrevented) {
            event.preventDefault();
        }
    }
    render() {
        return (h(Host, { key: '8d84ad8bd4a8f7298693be95382a9c1f154ddb22' }, h("div", { key: '3e8d41b1a01ba42136e7a44265dfd6afc50e6e3f', class: "alc-header-id" }, h("a", { key: '575c59311d6cefecaa8b7a637cd0d1595c1ad305', href: this.homeUrl, onClick: event => this.handleHome(event), class: "alc-header-id__link", "aria-label": `Página inicial do ${this.name}` }, h("img", { key: '5ecf3afd163951a1955126ab7837b7ec61d10888', class: "alc-header-id__logo", src: getAssetPath('assets/images/camara_logo.svg'), alt: "Logo da C\u00E2mara dos Deputados" }), h("div", { key: '286d929ccc651e7dcf180f34b476928df87574af', class: "alc-header-id__info" }, h("span", { key: '80a629b4ee68f3f01f3068853bac8fc8c2a8e4d6', class: "alc-header-id__name" }, this.name), this.description && (h("span", { key: '9d53cd1f1c5a4cb5741a150140ef476c74267ca5', class: "alc-header-id__description" }, this.description)))))));
    }
    static get is() { return "alc-header-id"; }
    static get originalStyleUrls() {
        return {
            "$": ["alc-header-id.css"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["alc-header-id.css"]
        };
    }
    static get assetsDirs() { return ["assets/images"]; }
    static get properties() {
        return {
            "name": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": true,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "O nome do sistema. Ser\u00E1 mostrado no cabe\u00E7alho, que deve estar presente em todas do sistema."
                },
                "getter": false,
                "setter": false,
                "attribute": "name",
                "reflect": true
            },
            "description": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Descri\u00E7\u00E3o do sistema. Usualmente, o nome do sistema (se usada a sigla em \"name\"), o nome de um m\u00F3dulo, ou um complemento ao nome do sistema."
                },
                "getter": false,
                "setter": false,
                "attribute": "description",
                "reflect": true
            },
            "homeUrl": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": true,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "URL para a p\u00E1gina inicial do sistema. O nome do sistema ser\u00E1 transformado em um link para essa URL."
                },
                "getter": false,
                "setter": false,
                "attribute": "home-url",
                "reflect": true
            }
        };
    }
    static get events() {
        return [{
                "method": "alcHome",
                "name": "alc-home",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Evento disparado quando o usu\u00E1rio aciona o link para a p\u00E1gina inicial.\nExecutar `preventDefault()` evita que a navega\u00E7\u00E3o para a p\u00E1gina inicial aconte\u00E7a."
                },
                "complexType": {
                    "original": "null",
                    "resolved": "null",
                    "references": {}
                }
            }];
    }
}
