// alcance/dist/components/alc-user.js
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { d as defineCustomElement$2 } from './alc-icon2.js';

const alcUserCss = ".alc-user{display:-ms-flexbox;display:flex}.alc-user__image{margin-right:1rem;height:3.5rem;width:3.5rem;border-radius:9999px;-o-object-fit:cover;object-fit:cover}.alc-user__image--icon{font-size:3.5rem;line-height:1}.alc-user__name{margin-top:0.5rem;font-size:1.125rem;line-height:1.5rem;font-weight:700}";

const AlcUser$1 = /*@__PURE__*/ proxyCustomElement(class AlcUser extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.alcLogout = createEvent(this, "alc-logout", 7);
        /**
         * Nome do usuário.
         */
        this.name = 'Usuário';
        /**
         * Número de matrícula ou ponto do usuário.
         */
        this.registrationNumber = '';
        /**
         * URL que efetua o logout do usuário.
         */
        this.logoutUrl = '#';
        /**
         * Imagem do usuário. Pode ser qualquer valor válido para o atributo `src` da tag `img` do HTML.
         */
        this.imgSrc = '';
    }
    async handleLogout(event) {
        const allowLogout = this.alcLogout.emit();
        if (allowLogout.defaultPrevented) {
            event.preventDefault();
        }
    }
    render() {
        return (h("div", { key: 'b880f04f7c16cd12bbff75f682e2027df7aaf0e0', class: "alc-user" }, this.imgSrc ? (h("img", { class: "alc-user__image", src: this.imgSrc, alt: "" })) : (h("alc-icon", { class: "alc-user__image alc-user__image--icon", name: "person-circle", label: "" })), h("div", { key: '8413b4f7e3df2f870b96dc7c46e3e37a62d84de8' }, h("div", { key: '6aee9461e654e08b0e47255c11df6ff4c92e61d0', class: "alc-user__name" }, this.name), h("div", { key: '344f61f96eacf268712084b4990384133d37e494' }, this.registrationNumber), h("slot", { key: '57b56ec0b6450763e1adab33ba6c69abb8ef9d15' }), h("div", { key: '9b10ea4a41c5bd4d0c66effdd0d4ca308ad03706' }, h("alc-icon", { key: '760138d17d0de1b12bf0757e7de19d99cc903059', name: "box-arrow-right", label: "" }), h("a", { key: '9733df1f115340fe20a43165ae8ab935d6ca91a1', class: "alc-link ml-2", href: this.logoutUrl, onClick: (event) => this.handleLogout(event) }, "Sair")))));
    }
    static get style() { return alcUserCss; }
}, [4, "alc-user", {
        "name": [513],
        "registrationNumber": [513, "registration-number"],
        "logoutUrl": [513, "logout-url"],
        "imgSrc": [513, "img-src"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["alc-user", "alc-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "alc-user":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, AlcUser$1);
            }
            break;
        case "alc-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const AlcUser = AlcUser$1;
const defineCustomElement = defineCustomElement$1;

export { AlcUser, defineCustomElement };
