// alcance/dist/components/alc-user-menu.js
import { proxyCustomElement, HTMLElement, createEvent, h, Host } from '@stencil/core/internal/client';
import { l as logger } from './logger.js';
import { d as defineCustomElement$9 } from './alc-dropdown2.js';
import { d as defineCustomElement$8 } from './alc-expander2.js';
import { d as defineCustomElement$7 } from './alc-icon2.js';
import { d as defineCustomElement$6 } from './alc-menu2.js';
import { d as defineCustomElement$5 } from './alc-menu-item2.js';
import { d as defineCustomElement$4 } from './alc-menu-item-theme2.js';
import { d as defineCustomElement$3 } from './alc-menu-link2.js';
import { d as defineCustomElement$2 } from './alc-popup2.js';

const alcUserMenuCss = "alc-user-menu{--avatar-size:2.5rem;--avatar-size-mobile:4rem;--border-width:2px}.alc-user-menu__image{margin-right:var(--alc-spacing-03);height:var(--avatar-size);width:var(--avatar-size);border-radius:9999px;border-width:calc(var(--border-width) * 1);--tw-border-opacity:1;border-color:rgb(0 104 125 / var(--tw-border-opacity));-o-object-fit:cover;object-fit:cover}.alc-user-menu__image:is([data-alc-theme=\"dark\"] *){--tw-border-opacity:1;border-color:rgb(79 158 153 / var(--tw-border-opacity))}.alc-user-menu__icon{margin-right:var(--alc-spacing-03);border-radius:9999px;border-width:calc(var(--border-width) * 1);--tw-border-opacity:1;border-color:rgb(0 104 125 / var(--tw-border-opacity));-o-object-fit:cover;object-fit:cover;font-size:calc(var(--avatar-size) - 2 * var(--border-width));line-height:1}.alc-user-menu__icon:is([data-alc-theme=\"dark\"] *){--tw-border-opacity:1;border-color:rgb(79 158 153 / var(--tw-border-opacity))}.alc-user-menu--mobile{display:-ms-flexbox;display:flex;-ms-flex-direction:column;flex-direction:column}.alc-user-menu--mobile__header{display:-ms-flexbox;display:flex;-ms-flex-align:center;align-items:center;padding-left:var(--alc-spacing-06);padding-right:var(--alc-spacing-09);padding-top:var(--alc-spacing-06)}.alc-user-menu--mobile__header .alc-user-menu__image{margin-right:var(--alc-spacing-05);height:var(--avatar-size-mobile);width:var(--avatar-size-mobile);border-style:none}.alc-user-menu--mobile__header .alc-user-menu__icon{background-color:var(--alc-color-layer-a);font-size:calc(var(--avatar-size-mobile) * 1);line-height:1}.alc-user-menu--mobile .alc-user-menu__name{font-size:1rem;line-height:1.5rem;font-weight:500}.alc-user-menu--desktop{display:-ms-flexbox;display:flex;-ms-flex-align:center;align-items:center;-ms-flex-pack:justify;justify-content:space-between}.alc-user-menu__info--external{margin-right:var(--alc-spacing-03);margin-left:auto;display:-ms-flexbox;display:flex;-ms-flex-direction:column;flex-direction:column;text-align:right}.alc-user-menu__info--internal{padding-left:var(--alc-spacing-05);padding-right:var(--alc-spacing-05);padding-bottom:var(--alc-spacing-03);padding-top:var(--alc-spacing-05);color:var(--alc-color-text-primary)}.alc-user-menu__info--internal:empty{padding:0px}.alc-user-menu__info{margin-right:var(--alc-spacing-03);display:-ms-flexbox;display:flex;-ms-flex-direction:column;flex-direction:column}.alc-user-menu__primary{-ms-flex-align:center;align-items:center;gap:var(--alc-spacing-02)}.alc-user-menu__name{font-size:1rem;line-height:var(--alc-font-line-height-base);font-weight:500}.alc-user-menu__reg{margin-left:var(--alc-spacing-02);font-weight:400}.alc-user-menu--mobile__reg{margin-left:var(--alc-spacing-02);font-weight:400}.alc-popup__content{max-width:16rem}.alc-user-menu__trigger{display:-ms-flexbox;display:flex;cursor:pointer;-ms-flex-align:center;align-items:center;border-style:none;background-color:transparent;padding:0px}.alc-user-menu__dropdown alc-popup{display:-ms-flexbox;display:flex}.alc-user-menu__secondary{font-size:0.875rem;line-height:1rem;font-weight:400;color:var(--alc-color-text-subtle)}.alc-user-menu__info--external .alc-user-menu__secondary{color:currentColor}.alc-user-menu__logout-icon{margin-right:var(--alc-spacing-03)}.alc-user-menu__info--separator{border-bottom-width:1px;border-color:var(--alc-color-border)}";

const AlcUserMenu$1 = /*@__PURE__*/ proxyCustomElement(class AlcUserMenu extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.alcLogout = createEvent(this, "alc-logout", 7);
        /**
         * Define a variação do menu de usuário. Geralmente, não é recomendado alterar esse valor.
         */
        this.variation = 'desktop';
        /**
         * Nome do usuário.
         */
        this.name = 'Usuário';
        /**
         * Número de matrícula ou ponto do usuário.
         */
        this.registrationNumber = '';
        /**
         * URL da imagem do usuário. Se não for fornecida, será exibido um ícone padrão.
         */
        this.imgSrc = '';
        this.dropdownOpen = false;
        this.handleLogout = (e) => {
            const { defaultPrevented } = this.alcLogout.emit();
            if (defaultPrevented) {
                e.preventDefault();
            }
        };
    }
    componentDidLoad() {
        if (this.variation === 'desktop') {
            this.externalContainer = this.host.querySelector('.alc-user-menu__info--external');
            this.internalContainer = this.host.querySelector('.alc-user-menu__info--internal');
            this.moveInfo();
        }
    }
    handleResize() {
        if (this.variation === 'desktop') {
            this.moveInfo();
        }
    }
    moveInfo() {
        const largura = window.innerWidth;
        // < 1200px é compacto, >= 1200px é expandido
        if (largura < 1200) {
            if (this.externalContainer && this.internalContainer && this.internalContainer.children.length === 0) {
                while (this.externalContainer.firstChild) {
                    this.internalContainer.appendChild(this.externalContainer.firstChild);
                    this.internalContainer.classList.add('alc-user-menu__info--separator');
                }
            }
        }
        else {
            if (this.externalContainer && this.internalContainer && this.externalContainer.children.length === 0) {
                while (this.internalContainer.firstChild) {
                    this.externalContainer.appendChild(this.internalContainer.firstChild);
                    this.internalContainer.classList.remove('alc-user-menu__info--separator');
                }
            }
        }
    }
    componentDidRender() {
        if (this.variation === 'mobile') {
            this.mobileMenuContainer.insertAdjacentElement('beforeend', this.menu);
            this.mobileSlotContainer.insertAdjacentElement('beforeend', this.defaultSlot);
        }
        else {
            this.desktopMenuContainer.insertAdjacentElement('beforeend', this.menu);
            this.desktopSlotContainer.insertAdjacentElement('beforeend', this.defaultSlot);
        }
    }
    renderDesktop() {
        return (h("div", { class: {
                "alc-user-menu--desktop": true,
                'hidden': this.variation === 'mobile'
            } }, h("div", { class: "alc-user-menu__info--external" }, h("div", null, h("div", { class: "alc-user-menu__primary" }, h("span", { class: "alc-user-menu__name" }, this.name), this.registrationNumber && h("span", { class: "alc-user-menu__reg" }, "(", this.registrationNumber, ")")), h("div", { class: "alc-user-menu__secondary", ref: el => this.desktopSlotContainer = el }))), h("alc-dropdown", { class: "alc-user-menu__dropdown" }, h("button", { slot: "trigger", class: "alc-user-menu__trigger", "aria-label": "Abrir menu do usu\u00E1rio" }, this.imgSrc ? (h("img", { class: "alc-user-menu__image", src: this.imgSrc, alt: this.name })) : (h("alc-icon", { class: "alc-user-menu__image alc-user-menu__icon", name: "person-fill", label: this.name })), h("alc-icon", { icon: "chevron-down", label: "" })), h("div", { class: "alc-user-menu__info--internal" }), h("div", { ref: el => this.desktopMenuContainer = el }))));
    }
    renderMobile() {
        return (h("div", { class: {
                "alc-user-menu--mobile": true,
                'hidden': this.variation === 'desktop'
            } }, h("div", { class: "alc-user-menu--mobile__header" }, this.imgSrc ? (h("img", { class: "alc-user-menu__image", src: this.imgSrc, alt: this.name })) : (h("alc-icon", { class: "alc-user-menu__image alc-user-menu__icon", name: "person-fill", label: this.name })), h("div", { class: "alc-user-menu--mobile__info" }, h("div", { class: "alc-user-menu__name" }, this.name, this.registrationNumber && h("span", { class: "alc-user-menu--mobile__reg" }, "(", this.registrationNumber, ")")), h("div", { class: "alc-user-menu__secondary", ref: el => this.mobileSlotContainer = el }))), h("alc-expander", { label: "Menu do usu\u00E1rio", open: false, "hide-label": "true" }, h("div", null, h("div", { ref: el => this.mobileMenuContainer = el })))));
    }
    render() {
        var _a;
        (_a = this.logoutUrl) !== null && _a !== void 0 ? _a : logger.report('logoutUrl', this.host.tagName.toLowerCase(), this.host);
        return (h(Host, { key: '1df7026fa5897691438a13aebfb38ad075f942f0' }, h("alc-menu", { key: '113f70bfb848564bf76a688356a458042e5e4cc0', ref: el => this.menu = el }, h("slot", { key: '13bdfda9dbb3a473380af99c6f47dde82f81f065', name: "actions" }), h("alc-menu-item-theme", { key: '5dcb6428c37d40843c29d77c5d3c164516646c0a' }), h("alc-menu-link", { key: 'dc29f0f8eec9b925e2f99330113f6b5a26d843b6' }, h("a", { key: '93b72f1a292ed1a45250aa7b515b40f2480c05a2', href: this.logoutUrl, onClick: this.handleLogout }, h("alc-icon", { key: '50c85a0422c3d2d87b340309917f12c0dedf4a37', name: "box-arrow-right", label: "", class: "alc-user-menu__logout-icon" }), "Sair"))), h("div", { key: '9b8f2386263e18f980625e276ba8eb72a152e28d', ref: el => this.defaultSlot = el }, h("slot", { key: '99a076179acea8e92cc4bf7f825593fc2dede355' })), [
            this.renderMobile(),
            this.renderDesktop()
        ]));
    }
    get host() { return this; }
    static get style() { return alcUserMenuCss; }
}, [4, "alc-user-menu", {
        "variation": [513],
        "name": [513],
        "registrationNumber": [513, "registration-number"],
        "logoutUrl": [513, "logout-url"],
        "imgSrc": [513, "img-src"],
        "dropdownOpen": [32]
    }, [[9, "resize", "handleResize"]]]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["alc-user-menu", "alc-dropdown", "alc-expander", "alc-icon", "alc-menu", "alc-menu-item", "alc-menu-item-theme", "alc-menu-link", "alc-popup"];
    components.forEach(tagName => { switch (tagName) {
        case "alc-user-menu":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, AlcUserMenu$1);
            }
            break;
        case "alc-dropdown":
            if (!customElements.get(tagName)) {
                defineCustomElement$9();
            }
            break;
        case "alc-expander":
            if (!customElements.get(tagName)) {
                defineCustomElement$8();
            }
            break;
        case "alc-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$7();
            }
            break;
        case "alc-menu":
            if (!customElements.get(tagName)) {
                defineCustomElement$6();
            }
            break;
        case "alc-menu-item":
            if (!customElements.get(tagName)) {
                defineCustomElement$5();
            }
            break;
        case "alc-menu-item-theme":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "alc-menu-link":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "alc-popup":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const AlcUserMenu = AlcUserMenu$1;
const defineCustomElement = defineCustomElement$1;

export { AlcUserMenu, defineCustomElement };
