// alcance/dist/components/alc-tab.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { l as logger } from './logger.js';
import { t as testAttributes } from './testAttributes.js';

const alcTabBaseCss = "";

const themeAlcTabCss = ".alc-tabs__tab:focus-visible{outline-style:solid;outline-width:2px;outline-offset:-8px;outline-color:var(--outline-color)}.alc-tabs__tab--content{border-color:var(--border-color);background-color:var(--bg-color);padding:0.75rem;border-width:var(--border-width);margin-top:calc(-1 * var(--border-width))}";

const AlcTab$1 = /*@__PURE__*/ proxyCustomElement(class AlcTab extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        /**
         * Rótulo da tab.
         */
        this.label = '';
        /**
         * Se, dentro da tab, o primeiro elemento com conteúdo significativo for focalizável, deve ser definido como `true`.
         * Caso contrário, deve ser mantido o valor padrão (`false`).
         */
        this.contentFocus = false;
    }
    render() {
        var _a;
        (_a = this.tab) !== null && _a !== void 0 ? _a : logger.report('tab', this.el.tagName.toLowerCase(), this.el);
        const getId = () => {
            let id;
            if (this.el.id) {
                id = this.el.id;
            }
            else {
                id = `alc-tab_${this.el.tab}`;
            }
            return id;
        };
        return (h(Host, null, h("div", Object.assign({ hidden: !this.selected, role: 'tabpanel', id: getId(), tabindex: this.contentFocus ? null : '0', class: 'alc-tabs__tab' }, testAttributes('data-test-tab')), h("div", { class: 'alc-tabs__tab--content' }, h("slot", null)))));
    }
    get el() { return this; }
    static get style() { return {
        base: alcTabBaseCss,
        theme: themeAlcTabCss
    }; }
}, [36, "alc-tab", {
        "label": [513],
        "tab": [513],
        "selected": [516],
        "contentFocus": [516, "content-focus"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["alc-tab"];
    components.forEach(tagName => { switch (tagName) {
        case "alc-tab":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, AlcTab$1);
            }
            break;
    } });
}

const AlcTab = AlcTab$1;
const defineCustomElement = defineCustomElement$1;

export { AlcTab, defineCustomElement };
