// alcance/dist/components/alc-breadcrumb.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { t as testAttributes } from './testAttributes.js';

const alcBreadcrumbCss = ".alc-breadcrumb{padding-top:0.25rem;padding-bottom:0.25rem}.alc-breadcrumb__list{display:-ms-flexbox;display:flex;-ms-flex-align:baseline;align-items:baseline}";

const AlcBreadcrumb$1 = /*@__PURE__*/ proxyCustomElement(class AlcBreadcrumb extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.checkCurrentPage = () => {
            const items = this.el.querySelectorAll('alc-breadcrumb-item');
            items.forEach((item, index) => {
                // O ultimo breadcrumb-item é a pagina atual
                if (index === items.length - 1) {
                    item.setAttribute('aria-current', 'page');
                }
                else {
                    item.removeAttribute('aria-current');
                }
            });
        };
    }
    componentWillLoad() {
        this.checkCurrentPage();
    }
    render() {
        return (h(Host, { key: '7beb7e14130eb314e5519c4f10d7dc407e2fc404' }, h("nav", Object.assign({ key: '4682ed361c08d53eaed18ad8526b6fdba581043d', "aria-label": "Breadcrumb", class: "alc-breadcrumb" }, testAttributes('data-test-nav')), h("div", Object.assign({ key: '85a7d7cd9c299f75bd12f0266577b9a7480d9cbe', class: "alc-breadcrumb__list", role: "list" }, testAttributes('data-test-list')), h("slot", { key: '7af5fa7fea926a5bd97f314a6fdac56e67b729fa' })))));
    }
    get el() { return this; }
    static get style() { return alcBreadcrumbCss; }
}, [4, "alc-breadcrumb"]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["alc-breadcrumb"];
    components.forEach(tagName => { switch (tagName) {
        case "alc-breadcrumb":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, AlcBreadcrumb$1);
            }
            break;
    } });
}

const AlcBreadcrumb = AlcBreadcrumb$1;
const defineCustomElement = defineCustomElement$1;

export { AlcBreadcrumb, defineCustomElement };
