// alcance/dist/components/alc-fieldset.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { g as getUniqueId } from './getUniqueId.js';
import { l as logger } from './logger.js';
import { t as testAttributes } from './testAttributes.js';

const alcFieldsetCss = "alc-fieldset{display:block}[data-alc-theme=\"dark\"] alc-fieldset{--border-color:var(--alc-color-border);--text-color:var(--alc-color-text-4);--text-color-error:var(--alc-color-error-45)}[data-alc-theme=\"light\"] alc-fieldset{--border-color:#76766a;--text-color:var(--alc-color-text-4);--text-color-error:var(--alc-color-error-50)}.alc-fieldset{border-radius:0.25rem;border-width:1px;border-style:solid;border-color:var(--border-color);padding:0.75rem}.alc-fieldset__legend{padding-left:0.375rem;padding-right:0.375rem;font-weight:600}.alc-fieldset__text-container{margin-top:1rem;display:-ms-flexbox;display:flex;-ms-flex-direction:column;flex-direction:column;gap:0.25rem}.alc-fieldset__text{font-size:0.875rem;line-height:1rem;color:var(--text-color)}.alc-fieldset__text--error{color:var(--text-color-error)}";

const AlcFieldset$1 = /*@__PURE__*/ proxyCustomElement(class AlcFieldset extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.describedBy = new Map();
        this.errorId = '';
        this.hintId = '';
        /**
         * Texto de ajuda.
        */
        this.hint = '';
        /**
         * Indica se o fieldset é obrigatório.
        */
        this.required = false;
        /**
         * Texto de mensagem de erro.
         */
        this.errorMsg = '';
    }
    watchErrorMsg(newValue) {
        logger.debug('watchErrorMsg', newValue);
        if (newValue) {
            this.el.classList.add('alc-form__input--error');
        }
        else {
            this.el.classList.remove('alc-form__input--error');
        }
    }
    componentWillLoad() {
        // Independentemente de qualquer coisa, reserva ids para esses elementos.
        this.errorId = getUniqueId();
        this.hintId = getUniqueId();
        if (this.el) {
            this.watchErrorMsg(this.errorMsg);
        }
    }
    componentWillRender() {
        if (this.errorMsg) {
            this.describedBy.set('error', this.errorId);
        }
        else {
            this.describedBy.delete('error');
        }
        if (this.hint) {
            this.describedBy.set('hint', this.hintId);
        }
        else {
            this.describedBy.delete('hint');
        }
    }
    render() {
        return (h(Host, { key: '13da7d1291538854017ab47348c086c484de42e5' }, h("fieldset", Object.assign({ key: '1bf206aaf951c0f43965e8512513e6f67a638dc7', class: "alc-fieldset", "aria-describedby": mapToString(this.describedBy) }, testAttributes('data-test-fieldset')), h("legend", Object.assign({ key: '087915de3b2ab15449223a4df975d4d8077c9d49', class: "alc-fieldset__legend" }, testAttributes('data-test-legend')), this.legend, this.required
            ?
                // aria-hidden porque o "required" do input já é suficiente para informar.
                h("small", { "aria-hidden": "true" }, " (obrigat\u00F3rio)")
            :
                null), h("slot", { key: '9fbadf05a4719285b0d4e76ed95270f5dbadfcf7' }), h("div", { key: 'b2ffb8687aa486a5d1c45ddecb52d9f30f31753b', class: "alc-fieldset__text-container" }, this.errorMsg
            ?
                h("small", Object.assign({ class: "alc-fieldset__text alc-fieldset__text--error", id: this.errorId }, testAttributes('data-test-error')), this.errorMsg)
            :
                null, this.hint
            ?
                h("small", Object.assign({ class: "alc-fieldset__text", id: this.hintId }, testAttributes('data-test-hint')), this.hint)
            :
                null))));
    }
    get el() { return this; }
    static get watchers() { return {
        "errorMsg": ["watchErrorMsg"]
    }; }
    static get style() { return alcFieldsetCss; }
}, [4, "alc-fieldset", {
        "legend": [513],
        "hint": [513],
        "required": [516],
        "errorMsg": [1537, "error-msg"]
    }, undefined, {
        "errorMsg": ["watchErrorMsg"]
    }]);
// get Map and transform in a string with the values separated by spaces. return null if Map is empty.
function mapToString(map) {
    if (map.size > 0) {
        return Array.from(map.values()).join(' ');
    }
    else {
        return null;
    }
}
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["alc-fieldset"];
    components.forEach(tagName => { switch (tagName) {
        case "alc-fieldset":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, AlcFieldset$1);
            }
            break;
    } });
}

const AlcFieldset = AlcFieldset$1;
const defineCustomElement = defineCustomElement$1;

export { AlcFieldset, defineCustomElement };
