// alcance/dist/esm/theme-BR2kXvnz.js
/*
  Controle do tema.
  Inspirado em "Dark mode Series' Articles": https://dev.to/ayc0/series/12983
 */
function loadUserPreference() {
    return localStorage.getItem('alc-theme') || 'system';
}
function saveUserPreference(theme) {
    localStorage.setItem('alc-theme', theme);
}
function removeUserPreference() {
    localStorage.removeItem('alc-theme');
}
function getAppliedTheme(userPreference) {
    // Usuário definiu sua preferência (light ou dark)
    if (userPreference === 'light') {
        return 'light';
    }
    if (userPreference === 'dark') {
        return 'dark';
    }
    // Usuário não definiu sua preferência aqui
    // mas indicou dark por meio do sistema operacional
    if (matchMedia('(prefers-color-scheme: dark)').matches) {
        return 'dark';
    }
    // Por default, light
    return 'light';
}
function setAppliedTheme(theme) {
    let themeTag = document.querySelector('html').dataset.alcThemeTag;
    themeTag = themeTag === undefined ? 'html' : themeTag;
    if (themeTag === '') {
        return;
    }
    const tags = document.querySelectorAll(themeTag);
    tags.forEach(tag => {
        if (tag instanceof HTMLElement) {
            tag.dataset['alcTheme'] = theme;
        }
    });
}
function getMode() {
    // A página informou que não quer usar o tema do Alcance
    // usando o atributo data-alc-no-theme
    // <html data-alc-no-theme>
    const noTheme = document.querySelector('html').dataset.alcNoTheme !== undefined;
    return noTheme ? 'base' : 'theme';
}

export { getMode as a, saveUserPreference as b, getAppliedTheme as g, loadUserPreference as l, removeUserPreference as r, setAppliedTheme as s };
